{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MRN Normalization\n",
    "\n",
    "1.  Loading data into a gene expression matrix.\n",
    "2.  Munging data\n",
    "3.  MRN Normalization\n",
    "4.  Quick shot at maching learning.\n",
    "\n",
    "\"Trimmed Mean of M-values\" (TMM) normalization, published by [Robinson and Oshlack](https://www.frontiersin.org/articles/10.3389/fgene.2016.00164/full#B16) is a widely used method of normalizing gene expression in scRNA data.  A variant called MRN (Median Ratio Normalization) is described by [Maza et al.](https://www.tandfonline.com/doi/full/10.4161/cib.25849), and may perform slightly better than TMM.  We carry out MRN normalization on the P1902 scRNA data, and use the results in a quick machine learning application."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import packages.  Put plots \"inline\" in the notebook.  \n",
    "\n",
    "import numpy as np  # For numerical computations.\n",
    "import pandas as pd  # Pandas for data analysis.\n",
    "import matplotlib.pyplot as plt  # For basic plotting.\n",
    "import seaborn as sns # For pretty visualization in Seaborn.  See https://seaborn.pydata.org/\n",
    "\n",
    "import os # Working with file directories, etc.\n",
    "\n",
    "from IPython.display import display # Pretty display of data frames.\n",
    "\n",
    "# Put plots inline rather than in a pop-up.\n",
    "%matplotlib inline\n",
    "\n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning) #got annoyed of these red boxes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Go to section 2, if you have already loaded the data and have the pickles."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1.  Loading the data\n",
    "\n",
    "Start at section 2, if the raw data has already been processed and pickled.\n",
    "\n",
    "In this section, we load the gene expression data and some metadata.  The sequencing data is stored in a 'merged_gene_counts.txt'.  The metadata is stored in a separate file called 'meta_data_marty_inVitro_feb8.csv'.  We use the metadata to select bulk-cells from  experiment P9855.  The first function loads the metadata."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_meta(experiment, filename = 'meta_data_marty_inVitro_feb8.csv', report=True, bulks = False):\n",
    "    df = pd.read_csv(filename, sep=',',\n",
    "                     index_col=2, header=0, low_memory=False)\n",
    "    df = df[ df['Project_ID'] == experiment] # Only cells from the experiment.\n",
    "    if bulks:\n",
    "        df = df[ df['Number_Of_Cells'] >= 25] # Include only bulks.\n",
    "    else:\n",
    "        df = df[ df['Number_Of_Cells'] == 1] # Get rid of bulks.\n",
    "    if report:\n",
    "        if bulks:\n",
    "            print('{} bulks found in experiment {}'.format(len(df), experiment))\n",
    "        else:\n",
    "            print('{} single cells found in experiment {}'.format(len(df), experiment))\n",
    "        clones = df['Clone_ID'].unique()\n",
    "        print('{} Clones: {}'.format(len(clones), ', '.join(clones)))\n",
    "        print('The first five rows of the dataframe are below')\n",
    "        display(df[:5])\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 single cells found in experiment P9855\n",
      "0 Clones: \n",
      "The first five rows of the dataframe are below\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Project_ID</th>\n",
       "      <th>Donor</th>\n",
       "      <th>Number_Of_Cells</th>\n",
       "      <th>Clone_ID</th>\n",
       "      <th>In_Vivo_Clone_ID</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Project_ID, Donor, Number_Of_Cells, Clone_ID, In_Vivo_Clone_ID]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "meta_df = get_meta('P9855', filename = 'meta_data_marty_inVitro_feb8.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "99 bulks found in experiment P9855\n",
      "53 Clones: P1_A03, P1_A05, P1_A12, P1_D09, P1_D12, P1_F10, P1_H01, P1_H10, P1_A08, P1_B10, P1_C11, P1_F02, P1_F04, P2_A04, P2_B11, P2_D12, P2_F01, P2_G01, P2_G09, P2_H01, P2_H04, P2_H05, P2_A05, P2_C12, P1_A03_ATAC_wellA1_rep1_1000, P1_A05_ATAC_wellA3_705, P1_D09_ATAC_wellA5_410, P1_F10_ATAC_wellA8_1000, P1_H01_ATAC_wellA9_rep1_1000, P1_H01_ATAC_wellA10_rep2_1000, P1_H10_ATAC_wellA11_rep1_1000, P1_A08_ATAC_wellB1_350, P1_A03_ATAC_wellA2_rep2_1000, P1_A12_ATAC_wellA4_506, P1_D12_ATAC_wellA6_1000, P1_H10_ATAC_wellA12_rep2_1000, P1_B10_ATAC_wellB2_269, P1_C11_wellB3_500, P1_F02_wellB4_500, P1_F04_wellB5_500, P2_A04_wellC1_1000_rep1, P2_A04_wellC2_1000_rep2, P2_B11_wellC3_700, P2_H05_wellC5_1000, P2_F01_wellC6_1000, P2_D12_wellC7_1000, P2_G01_wellC8_1000, P2_G09_wellC9_918, P2_H01_wellC10_1000, P2_H04_wellC11_1000_rep1, P2_H04_wellC12_1000_rep2, P2_H05_wellD1_1000, P2_A05_wellD2_1000\n",
      "The first five rows of the dataframe are below\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Project_ID</th>\n",
       "      <th>Donor</th>\n",
       "      <th>Number_Of_Cells</th>\n",
       "      <th>Clone_ID</th>\n",
       "      <th>In_Vivo_Clone_ID</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P9855_2001</th>\n",
       "      <td>P9855</td>\n",
       "      <td>YFV2003</td>\n",
       "      <td>25</td>\n",
       "      <td>P1_A03</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P9855_2002</th>\n",
       "      <td>P9855</td>\n",
       "      <td>YFV2003</td>\n",
       "      <td>25</td>\n",
       "      <td>P1_A03</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P9855_2003</th>\n",
       "      <td>P9855</td>\n",
       "      <td>YFV2003</td>\n",
       "      <td>25</td>\n",
       "      <td>P1_A03</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P9855_2004</th>\n",
       "      <td>P9855</td>\n",
       "      <td>YFV2003</td>\n",
       "      <td>25</td>\n",
       "      <td>P1_A05</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P9855_2005</th>\n",
       "      <td>P9855</td>\n",
       "      <td>YFV2003</td>\n",
       "      <td>25</td>\n",
       "      <td>P1_A05</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Project_ID    Donor  Number_Of_Cells Clone_ID  In_Vivo_Clone_ID\n",
       "Sample_ID                                                                 \n",
       "P9855_2001      P9855  YFV2003               25   P1_A03               NaN\n",
       "P9855_2002      P9855  YFV2003               25   P1_A03               NaN\n",
       "P9855_2003      P9855  YFV2003               25   P1_A03               NaN\n",
       "P9855_2004      P9855  YFV2003               25   P1_A05               NaN\n",
       "P9855_2005      P9855  YFV2003               25   P1_A05               NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "meta_bulks = get_meta('P9855', filename = 'meta_data_marty_inVitro_feb8.csv', bulks=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Project_ID</th>\n",
       "      <th>Donor</th>\n",
       "      <th>Number_Of_Cells</th>\n",
       "      <th>Clone_ID</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P9855_2001</th>\n",
       "      <td>P9855</td>\n",
       "      <td>YFV2003</td>\n",
       "      <td>25</td>\n",
       "      <td>A03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P9855_2002</th>\n",
       "      <td>P9855</td>\n",
       "      <td>YFV2003</td>\n",
       "      <td>25</td>\n",
       "      <td>A03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P9855_2003</th>\n",
       "      <td>P9855</td>\n",
       "      <td>YFV2003</td>\n",
       "      <td>25</td>\n",
       "      <td>A03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P9855_2004</th>\n",
       "      <td>P9855</td>\n",
       "      <td>YFV2003</td>\n",
       "      <td>25</td>\n",
       "      <td>A05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P9855_2005</th>\n",
       "      <td>P9855</td>\n",
       "      <td>YFV2003</td>\n",
       "      <td>25</td>\n",
       "      <td>A05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Project_ID    Donor  Number_Of_Cells Clone_ID\n",
       "Sample_ID                                               \n",
       "P9855_2001      P9855  YFV2003               25      A03\n",
       "P9855_2002      P9855  YFV2003               25      A03\n",
       "P9855_2003      P9855  YFV2003               25      A03\n",
       "P9855_2004      P9855  YFV2003               25      A05\n",
       "P9855_2005      P9855  YFV2003               25      A05"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_bulks.drop(meta_bulks.index[[range(69,99)]], inplace=True) # drop atac data... and c12 while we are cutting\n",
    "                                                                # the tail off of the dataframe \n",
    "meta_bulks.drop(['In_Vivo_Clone_ID'],axis=1, inplace=True)\n",
    "meta_bulks['Clone_ID'] = meta_bulks['Clone_ID'].str[-3:] # cleaning up clone id names\n",
    "meta_bulks.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(69, 4)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_bulks.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['A03', 'A05', 'A12', 'D09', 'D12', 'F10', 'H01', 'H10', 'A08',\n",
       "       'B10', 'C11', 'F02', 'F04', 'A04', 'B11', 'F01', 'G01', 'G09',\n",
       "       'H04', 'H05'], dtype=object)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_bulks['Clone_ID'].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The sequencing data (in merged_gene_counts.txt) contains a separate row for each  gene.  The name of each column contains list of filenames (I'm guessing input for star), we will clean up column names."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>P9855_2019</th>\n",
       "      <th>P9855_2054</th>\n",
       "      <th>P9855_2072</th>\n",
       "      <th>P9855_2025</th>\n",
       "      <th>P9855_2076</th>\n",
       "      <th>P9855_2068</th>\n",
       "      <th>P9855_2006</th>\n",
       "      <th>P9855_2078</th>\n",
       "      <th>P9855_2042</th>\n",
       "      <th>...</th>\n",
       "      <th>P9855_2009</th>\n",
       "      <th>P9855_2016</th>\n",
       "      <th>P9855_2061</th>\n",
       "      <th>P9855_2015</th>\n",
       "      <th>P9855_2062</th>\n",
       "      <th>P9855_2012</th>\n",
       "      <th>P9855_2080</th>\n",
       "      <th>P9855_2070</th>\n",
       "      <th>P9855_2030</th>\n",
       "      <th>P9855_2031</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>MIR1302-10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613</th>\n",
       "      <td>FAM138A</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268020</th>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 74 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  gene_name  P9855_2019  P9855_2054  P9855_2072  P9855_2025  \\\n",
       "Geneid                                                                        \n",
       "ENSG00000223972     DDX11L1           0           0           0           0   \n",
       "ENSG00000227232      WASH7P           0           0           0           0   \n",
       "ENSG00000243485  MIR1302-10           0           0           0           0   \n",
       "ENSG00000237613     FAM138A           0           0           0           0   \n",
       "ENSG00000268020      OR4G4P           0           0           0           0   \n",
       "\n",
       "                 P9855_2076  P9855_2068  P9855_2006  P9855_2078  P9855_2042  \\\n",
       "Geneid                                                                        \n",
       "ENSG00000223972           0           0           0           0           0   \n",
       "ENSG00000227232           0           0           0           0           0   \n",
       "ENSG00000243485           0           0           0           0           0   \n",
       "ENSG00000237613           0           0           0           0           0   \n",
       "ENSG00000268020           0           0           0           0           0   \n",
       "\n",
       "                 ...  P9855_2009  P9855_2016  P9855_2061  P9855_2015  \\\n",
       "Geneid           ...                                                   \n",
       "ENSG00000223972  ...           0           0           0           0   \n",
       "ENSG00000227232  ...           0           0           0           0   \n",
       "ENSG00000243485  ...           0           0           0           0   \n",
       "ENSG00000237613  ...           0           0           0           0   \n",
       "ENSG00000268020  ...           0           0           0           0   \n",
       "\n",
       "                 P9855_2062  P9855_2012  P9855_2080  P9855_2070  P9855_2030  \\\n",
       "Geneid                                                                        \n",
       "ENSG00000223972           0           0           0           0           0   \n",
       "ENSG00000227232           0           0           0           0           0   \n",
       "ENSG00000243485           0           0           0           0           0   \n",
       "ENSG00000237613           0           0           0           0           0   \n",
       "ENSG00000268020           0           0           0           0           0   \n",
       "\n",
       "                 P9855_2031  \n",
       "Geneid                       \n",
       "ENSG00000223972           0  \n",
       "ENSG00000227232           0  \n",
       "ENSG00000243485           0  \n",
       "ENSG00000237613           0  \n",
       "ENSG00000268020           0  \n",
       "\n",
       "[5 rows x 74 columns]"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_merged_genes = pd.read_csv('merged_gene_counts.txt', sep='\\t', index_col=0)\n",
    "read_merged_genes.columns = read_merged_genes.columns.str[:10] # 10 characters in P9855_20**\n",
    "#read_merged_genes.drop(['gene_name'],axis=1, inplace=True)\n",
    "#read_merged_genes = read_merged_genes.transpose() # tanspose dataframe for later analysis infrastructure\n",
    "#read_merged_genes.index.names = ['Sample_ID']\n",
    "read_merged_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "read_merged_genes.to_pickle('P9855_rawreads.pkl') # Save file as a pickle.\n",
    "meta_bulks.to_pickle('P9855_meta.pkl') # Pickle the metadata too."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The next function loads an entire *list* of cells, and places their gene expression data into the rows of a matrix.  The rows are indexed by the cell names, and the columns by genes.  The data is the gene expression, as raw number of reads.  This may take a little while, so we give progress updates every 10 cells."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2.  Munging data\n",
    "\n",
    "Start here if you have the pickles!  We filter the data a bit, before normalization downstream."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "bulks_raw = pd.read_pickle('P9855_rawreads.pkl') # Load bulks expression matrix from a pickle.\n",
    "meta_df = pd.read_pickle('P9855_meta.pkl') # Load metadata from a pickle."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = list(bulks_raw.columns)  # The names of the genes. \n",
    "bulks = list(bulks_raw.index) # The names of the bulks.\n",
    "clones = sorted(list(meta_bulks.Clone_ID.unique())) # The names of the clones."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Removing TCRs and rarely-expressed genes\n",
    "\n",
    "T cells have special genetically rearranged receptors called TCRs.  These are made of segments called TRBV9, TRBJ2-4, TRAV12-2, TRAJ14, etc.  Bascally any gene that is called these letters followed by a number -- TRBV, TRBJ, TRAV, TRAJ -- is part of this receptor and they are defined as being clonal.  Therefore we exclude these genes since we want to find more interesting similarities within clonal populations.\n",
    "\n",
    "The following loads a list of genes to be excluded from the data for later analysis.  The excluded genes should be given in a csv file with *one* column.  No row labels should be given.  The first row should be a descriptive header, like \"Genes to exclude.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [],
   "source": [
    "exc_filename = 'TRgenes.csv'  # CHANGE this if needed.  I added TRAC and TRDV3 as requested.\n",
    "exc_df = pd.read_csv(exc_filename, sep=',', header=0)\n",
    "exclude_genes = exc_df.iloc[:,0].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "ensembl_hgnc_meta_filename = 'All_Gene_Info_Metadata.csv'\n",
    "ensembl_hgnc_meta_df = pd.read_csv(ensembl_hgnc_meta_filename, sep=',', header=0)\n",
    "ensembl_gene_id = ensembl_hgnc_meta_df.iloc[:,:2].to_dict(orient='records')\n",
    "#exclude_ensembl = [ensembl_id for gene in ensembl_gene_id if gene in exclude_genes]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'ensembl_gene_id': 'ENSG00000261657', 'hgnc_symbol': 'SLC25A26'},\n",
       " {'ensembl_gene_id': 'ENSG00000261657', 'hgnc_symbol': 'SLC25A26'},\n",
       " {'ensembl_gene_id': 'ENSG00000261657', 'hgnc_symbol': 'SLC25A26'},\n",
       " {'ensembl_gene_id': 'ENSG00000223116', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000233440', 'hgnc_symbol': 'HMGA1P6'},\n",
       " {'ensembl_gene_id': 'ENSG00000207157', 'hgnc_symbol': 'RNY3P4'},\n",
       " {'ensembl_gene_id': 'ENSG00000229483', 'hgnc_symbol': 'LINC00362'},\n",
       " {'ensembl_gene_id': 'ENSG00000252952', 'hgnc_symbol': 'RNU6-58P'},\n",
       " {'ensembl_gene_id': 'ENSG00000235205', 'hgnc_symbol': 'TATDN2P3'},\n",
       " {'ensembl_gene_id': 'ENSG00000232849', 'hgnc_symbol': 'LINC00363'},\n",
       " {'ensembl_gene_id': 'ENSG00000236803', 'hgnc_symbol': 'SDAD1P4'},\n",
       " {'ensembl_gene_id': 'ENSG00000232163', 'hgnc_symbol': 'RPLP1P13'},\n",
       " {'ensembl_gene_id': 'ENSG00000252365', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000229558', 'hgnc_symbol': 'SACS-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000235784', 'hgnc_symbol': 'HNRNPA1P29'},\n",
       " {'ensembl_gene_id': 'ENSG00000224394', 'hgnc_symbol': 'GPC6-AS2'},\n",
       " {'ensembl_gene_id': 'ENSG00000212559', 'hgnc_symbol': 'RNA5SP35'},\n",
       " {'ensembl_gene_id': 'ENSG00000232977', 'hgnc_symbol': 'LINC00327'},\n",
       " {'ensembl_gene_id': 'ENSG00000232977', 'hgnc_symbol': 'LINC00327'},\n",
       " {'ensembl_gene_id': 'ENSG00000236520', 'hgnc_symbol': 'GPC6-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000227893', 'hgnc_symbol': 'LINC00352'},\n",
       " {'ensembl_gene_id': 'ENSG00000212057', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000222129', 'hgnc_symbol': 'RNA5SP36'},\n",
       " {'ensembl_gene_id': 'ENSG00000232102', 'hgnc_symbol': 'MTCO3P2'},\n",
       " {'ensembl_gene_id': 'ENSG00000242037', 'hgnc_symbol': 'RN7SL585P'},\n",
       " {'ensembl_gene_id': 'ENSG00000238230', 'hgnc_symbol': 'LINC00391'},\n",
       " {'ensembl_gene_id': 'ENSG00000227640', 'hgnc_symbol': 'SOX21-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000205861', 'hgnc_symbol': 'C1QTNF9B-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000205861', 'hgnc_symbol': 'C1QTNF9B-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000205861', 'hgnc_symbol': 'C1QTNF9B-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000235768', 'hgnc_symbol': 'BRD7P5'},\n",
       " {'ensembl_gene_id': 'ENSG00000237924', 'hgnc_symbol': 'RPL21P112'},\n",
       " {'ensembl_gene_id': 'ENSG00000260962', 'hgnc_symbol': 'LINC00557'},\n",
       " {'ensembl_gene_id': 'ENSG00000252335', 'hgnc_symbol': 'RNU6-62P'},\n",
       " {'ensembl_gene_id': 'ENSG00000238463', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000223298', 'hgnc_symbol': 'RNY3P8'},\n",
       " {'ensembl_gene_id': 'ENSG00000196593', 'hgnc_symbol': 'ANKRD20A19P'},\n",
       " {'ensembl_gene_id': 'ENSG00000196593', 'hgnc_symbol': 'ANKRD20A19P'},\n",
       " {'ensembl_gene_id': 'ENSG00000196593', 'hgnc_symbol': 'ANKRD20A19P'},\n",
       " {'ensembl_gene_id': 'ENSG00000200211', 'hgnc_symbol': 'RNY4P27'},\n",
       " {'ensembl_gene_id': 'ENSG00000271492', 'hgnc_symbol': 'MEMO1P5'},\n",
       " {'ensembl_gene_id': 'ENSG00000223392', 'hgnc_symbol': 'CLDN10-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000228741', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000247400', 'hgnc_symbol': 'DNAJC3-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000247400', 'hgnc_symbol': 'DNAJC3-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000225674', 'hgnc_symbol': 'IPO7P2'},\n",
       " {'ensembl_gene_id': 'ENSG00000229622', 'hgnc_symbol': 'MTND5P2'},\n",
       " {'ensembl_gene_id': 'ENSG00000252695', 'hgnc_symbol': 'MIR2276'},\n",
       " {'ensembl_gene_id': 'ENSG00000227213', 'hgnc_symbol': 'SPATA13-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000227714', 'hgnc_symbol': 'MTND6P18'},\n",
       " {'ensembl_gene_id': 'ENSG00000240868', 'hgnc_symbol': 'C1QTNF9-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000232373', 'hgnc_symbol': 'MTCYBP3'},\n",
       " {'ensembl_gene_id': 'ENSG00000234627', 'hgnc_symbol': 'NUS1P3'},\n",
       " {'ensembl_gene_id': 'ENSG00000234627', 'hgnc_symbol': 'NUS1P3'},\n",
       " {'ensembl_gene_id': 'ENSG00000251901', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000261498', 'hgnc_symbol': 'LINC00566'},\n",
       " {'ensembl_gene_id': 'ENSG00000226977', 'hgnc_symbol': 'HMGN1P24'},\n",
       " {'ensembl_gene_id': 'ENSG00000233545', 'hgnc_symbol': 'CYCSP33'},\n",
       " {'ensembl_gene_id': 'ENSG00000242614', 'hgnc_symbol': 'RN7SL164P'},\n",
       " {'ensembl_gene_id': 'ENSG00000233273', 'hgnc_symbol': 'AMMECR1LP1'},\n",
       " {'ensembl_gene_id': 'ENSG00000205822', 'hgnc_symbol': 'TPTE2P6'},\n",
       " {'ensembl_gene_id': 'ENSG00000205822', 'hgnc_symbol': 'TPTE2P6'},\n",
       " {'ensembl_gene_id': 'ENSG00000205822', 'hgnc_symbol': 'TPTE2P6'},\n",
       " {'ensembl_gene_id': 'ENSG00000235137', 'hgnc_symbol': 'HSP90AB6P'},\n",
       " {'ensembl_gene_id': 'ENSG00000235137', 'hgnc_symbol': 'HSP90AB6P'},\n",
       " {'ensembl_gene_id': 'ENSG00000243300', 'hgnc_symbol': 'LINC00359'},\n",
       " {'ensembl_gene_id': 'ENSG00000243300', 'hgnc_symbol': 'LINC00359'},\n",
       " {'ensembl_gene_id': 'ENSG00000266216', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000244615', 'hgnc_symbol': 'PSPC1P2'},\n",
       " {'ensembl_gene_id': 'ENSG00000234267', 'hgnc_symbol': 'TULP3P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000243008', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000272329', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000272329', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000211508', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000222472', 'hgnc_symbol': 'RN7SKP7'},\n",
       " {'ensembl_gene_id': 'ENSG00000207474', 'hgnc_symbol': 'RNY1P7'},\n",
       " {'ensembl_gene_id': 'ENSG00000252154', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000225404', 'hgnc_symbol': 'RPL26P34'},\n",
       " {'ensembl_gene_id': 'ENSG00000238522', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000233124', 'hgnc_symbol': 'LINC00456'},\n",
       " {'ensembl_gene_id': 'ENSG00000231417', 'hgnc_symbol': 'IRX1P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000202290', 'hgnc_symbol': 'RNA5SP37'},\n",
       " {'ensembl_gene_id': 'ENSG00000271625', 'hgnc_symbol': 'PSMA6P4'},\n",
       " {'ensembl_gene_id': 'ENSG00000226695', 'hgnc_symbol': 'ANKRD20A10P'},\n",
       " {'ensembl_gene_id': 'ENSG00000226134', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000238407', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000233291', 'hgnc_symbol': 'RPL7AP61'},\n",
       " {'ensembl_gene_id': 'ENSG00000270725', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000253771', 'hgnc_symbol': 'TPTE2P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000253771', 'hgnc_symbol': 'TPTE2P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000253771', 'hgnc_symbol': 'TPTE2P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000253771', 'hgnc_symbol': 'TPTE2P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000232217', 'hgnc_symbol': 'FTLP8'},\n",
       " {'ensembl_gene_id': 'ENSG00000222969', 'hgnc_symbol': 'RN7SKP8'},\n",
       " {'ensembl_gene_id': 'ENSG00000263399', 'hgnc_symbol': 'MIR3170'},\n",
       " {'ensembl_gene_id': 'ENSG00000269189', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000269599', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000251299', 'hgnc_symbol': 'SLC25A15P3'},\n",
       " {'ensembl_gene_id': 'ENSG00000231194', 'hgnc_symbol': 'FARP1-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000271437', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000232858', 'hgnc_symbol': 'RPL34P27'},\n",
       " {'ensembl_gene_id': 'ENSG00000224418', 'hgnc_symbol': 'STK24-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000256305', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000217686', 'hgnc_symbol': 'NUS1P4'},\n",
       " {'ensembl_gene_id': 'ENSG00000234056', 'hgnc_symbol': 'AMER2-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000233549', 'hgnc_symbol': 'CYCSP35'},\n",
       " {'ensembl_gene_id': 'ENSG00000238169', 'hgnc_symbol': 'LINC01053'},\n",
       " {'ensembl_gene_id': 'ENSG00000233662', 'hgnc_symbol': 'CALM2P4'},\n",
       " {'ensembl_gene_id': 'ENSG00000225105', 'hgnc_symbol': 'LINC01076'},\n",
       " {'ensembl_gene_id': 'ENSG00000243366', 'hgnc_symbol': 'RN7SL60P'},\n",
       " {'ensembl_gene_id': 'ENSG00000235315', 'hgnc_symbol': 'RPL23AP69'},\n",
       " {'ensembl_gene_id': 'ENSG00000229918', 'hgnc_symbol': 'DOCK9-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000261530', 'hgnc_symbol': 'BTC'},\n",
       " {'ensembl_gene_id': 'ENSG00000261530', 'hgnc_symbol': 'BTC'},\n",
       " {'ensembl_gene_id': 'ENSG00000221324', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000260509', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000231981', 'hgnc_symbol': 'RPL7L1P12'},\n",
       " {'ensembl_gene_id': 'ENSG00000234152', 'hgnc_symbol': 'TCEB2P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000265546', 'hgnc_symbol': 'RN7SL289P'},\n",
       " {'ensembl_gene_id': 'ENSG00000260666', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000207298', 'hgnc_symbol': 'RNU6-83P'},\n",
       " {'ensembl_gene_id': 'ENSG00000199767', 'hgnc_symbol': 'RNU6-78P'},\n",
       " {'ensembl_gene_id': 'ENSG00000260992', 'hgnc_symbol': 'DOCK9-AS2'},\n",
       " {'ensembl_gene_id': 'ENSG00000180803', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000230871', 'hgnc_symbol': 'RPS6P23'},\n",
       " {'ensembl_gene_id': 'ENSG00000242381', 'hgnc_symbol': 'RN7SL741P'},\n",
       " {'ensembl_gene_id': 'ENSG00000260334', 'hgnc_symbol': 'GAST'},\n",
       " {'ensembl_gene_id': 'ENSG00000223158', 'hgnc_symbol': 'RNY1P3'},\n",
       " {'ensembl_gene_id': 'ENSG00000231983', 'hgnc_symbol': 'LINC00415'},\n",
       " {'ensembl_gene_id': 'ENSG00000234005', 'hgnc_symbol': 'GAPDHP22'},\n",
       " {'ensembl_gene_id': 'ENSG00000228889', 'hgnc_symbol': 'UBAC2-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000233963', 'hgnc_symbol': 'ATP8A2P3'},\n",
       " {'ensembl_gene_id': 'ENSG00000201793', 'hgnc_symbol': 'RN7SKP9'},\n",
       " {'ensembl_gene_id': 'ENSG00000218502', 'hgnc_symbol': 'H2AFZP3'},\n",
       " {'ensembl_gene_id': 'ENSG00000207719', 'hgnc_symbol': 'MIR623'},\n",
       " {'ensembl_gene_id': 'ENSG00000234706', 'hgnc_symbol': 'PRUNEP1'},\n",
       " {'ensembl_gene_id': 'ENSG00000228808', 'hgnc_symbol': 'HMGB3P4'},\n",
       " {'ensembl_gene_id': 'ENSG00000271559', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000238241', 'hgnc_symbol': 'CCR12P'},\n",
       " {'ensembl_gene_id': 'ENSG00000227882', 'hgnc_symbol': 'PRKRIRP6'},\n",
       " {'ensembl_gene_id': 'ENSG00000266207', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000203441', 'hgnc_symbol': 'LINC00449'},\n",
       " {'ensembl_gene_id': 'ENSG00000234031', 'hgnc_symbol': 'RPS3AP44'},\n",
       " {'ensembl_gene_id': 'ENSG00000212197', 'hgnc_symbol': 'RNY3P6'},\n",
       " {'ensembl_gene_id': 'ENSG00000234168', 'hgnc_symbol': 'LINC01039'},\n",
       " {'ensembl_gene_id': 'ENSG00000237001', 'hgnc_symbol': 'WASF3-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000234239', 'hgnc_symbol': 'CFL1P8'},\n",
       " {'ensembl_gene_id': 'ENSG00000230256', 'hgnc_symbol': 'FGFR1OP2P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000223782', 'hgnc_symbol': 'RPS21P8'},\n",
       " {'ensembl_gene_id': 'ENSG00000263615', 'hgnc_symbol': 'MIR4306'},\n",
       " {'ensembl_gene_id': 'ENSG00000218198', 'hgnc_symbol': 'RPS20P32'},\n",
       " {'ensembl_gene_id': 'ENSG00000218198', 'hgnc_symbol': 'RPS20P32'},\n",
       " {'ensembl_gene_id': 'ENSG00000227659', 'hgnc_symbol': 'CLYBL-AS2'},\n",
       " {'ensembl_gene_id': 'ENSG00000232162', 'hgnc_symbol': 'USP12-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000234303', 'hgnc_symbol': 'CLYBL-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000230641', 'hgnc_symbol': 'USP12-AS2'},\n",
       " {'ensembl_gene_id': 'ENSG00000234772', 'hgnc_symbol': 'LINC00412'},\n",
       " {'ensembl_gene_id': 'ENSG00000201245', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000238305', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000207500', 'hgnc_symbol': 'SNORD102'},\n",
       " {'ensembl_gene_id': 'ENSG00000260738', 'hgnc_symbol': 'LINC00554'},\n",
       " {'ensembl_gene_id': 'ENSG00000207051', 'hgnc_symbol': 'SNORA27'},\n",
       " {'ensembl_gene_id': 'ENSG00000221756', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000225999', 'hgnc_symbol': 'NDUFA12P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000252247', 'hgnc_symbol': 'RNU6-70P'},\n",
       " {'ensembl_gene_id': 'ENSG00000229609', 'hgnc_symbol': 'LINC01079'},\n",
       " {'ensembl_gene_id': 'ENSG00000236554', 'hgnc_symbol': 'ASNSP3'},\n",
       " {'ensembl_gene_id': 'ENSG00000201242', 'hgnc_symbol': 'RNY1P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000203602', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000252499', 'hgnc_symbol': 'RNU6-63P'},\n",
       " {'ensembl_gene_id': 'ENSG00000228501', 'hgnc_symbol': 'RPL15P18'},\n",
       " {'ensembl_gene_id': 'ENSG00000226703', 'hgnc_symbol': 'NPM1P4'},\n",
       " {'ensembl_gene_id': 'ENSG00000261125', 'hgnc_symbol': 'HAP1'},\n",
       " {'ensembl_gene_id': 'ENSG00000261125', 'hgnc_symbol': 'HAP1'},\n",
       " {'ensembl_gene_id': 'ENSG00000261125', 'hgnc_symbol': 'HAP1'},\n",
       " {'ensembl_gene_id': 'ENSG00000261125', 'hgnc_symbol': 'HAP1'},\n",
       " {'ensembl_gene_id': 'ENSG00000270632', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000207330', 'hgnc_symbol': 'RNU6-73P'},\n",
       " {'ensembl_gene_id': 'ENSG00000234650', 'hgnc_symbol': 'PCCA-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000234650', 'hgnc_symbol': 'PCCA-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000234650', 'hgnc_symbol': 'PCCA-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000247381', 'hgnc_symbol': 'PDX1-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000260704', 'hgnc_symbol': 'LINC00543'},\n",
       " {'ensembl_gene_id': 'ENSG00000242360', 'hgnc_symbol': 'RN7SL272P'},\n",
       " {'ensembl_gene_id': 'ENSG00000224356', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000224356', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000271026', 'hgnc_symbol': 'KATNBL1P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000235115', 'hgnc_symbol': 'CHCHD2P8'},\n",
       " {'ensembl_gene_id': 'ENSG00000235115', 'hgnc_symbol': 'CHCHD2P8'},\n",
       " {'ensembl_gene_id': 'ENSG00000234354', 'hgnc_symbol': 'RPS26P47'},\n",
       " {'ensembl_gene_id': 'ENSG00000234354', 'hgnc_symbol': 'RPS26P47'},\n",
       " {'ensembl_gene_id': 'ENSG00000261485', 'hgnc_symbol': 'PAN3-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000237082', 'hgnc_symbol': 'COX5BP6'},\n",
       " {'ensembl_gene_id': 'ENSG00000232587', 'hgnc_symbol': 'EEF1A1P3'},\n",
       " {'ensembl_gene_id': 'ENSG00000200840', 'hgnc_symbol': 'RNU6-82P'},\n",
       " {'ensembl_gene_id': 'ENSG00000233009', 'hgnc_symbol': 'NALCN-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000235472', 'hgnc_symbol': 'EIF4A1P7'},\n",
       " {'ensembl_gene_id': 'ENSG00000202237', 'hgnc_symbol': 'RNU6-53P'},\n",
       " {'ensembl_gene_id': 'ENSG00000214051', 'hgnc_symbol': 'ARF4P3'},\n",
       " {'ensembl_gene_id': 'ENSG00000233588', 'hgnc_symbol': 'CYP51A1P2'},\n",
       " {'ensembl_gene_id': 'ENSG00000229599', 'hgnc_symbol': 'LINC00411'},\n",
       " {'ensembl_gene_id': 'ENSG00000236339', 'hgnc_symbol': 'POM121L13P'},\n",
       " {'ensembl_gene_id': 'ENSG00000272349', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000221670', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000201155', 'hgnc_symbol': 'RNU1-24P'},\n",
       " {'ensembl_gene_id': 'ENSG00000236758', 'hgnc_symbol': 'MTUS2-AS2'},\n",
       " {'ensembl_gene_id': 'ENSG00000215399', 'hgnc_symbol': 'HMGB3P7'},\n",
       " {'ensembl_gene_id': 'ENSG00000265164', 'hgnc_symbol': 'MIR2681'},\n",
       " {'ensembl_gene_id': 'ENSG00000223460', 'hgnc_symbol': 'GAPDHP69'},\n",
       " {'ensembl_gene_id': 'ENSG00000229975', 'hgnc_symbol': 'LIPT1P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000270680', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000201690', 'hgnc_symbol': 'RNY1P2'},\n",
       " {'ensembl_gene_id': 'ENSG00000179141', 'hgnc_symbol': 'MTUS2-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000179141', 'hgnc_symbol': 'MTUS2-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000264482', 'hgnc_symbol': 'MIR4705'},\n",
       " {'ensembl_gene_id': 'ENSG00000236860', 'hgnc_symbol': 'RPL39P29'},\n",
       " {'ensembl_gene_id': 'ENSG00000236258', 'hgnc_symbol': 'TIMM8BP1'},\n",
       " {'ensembl_gene_id': 'ENSG00000224329', 'hgnc_symbol': 'LINC00297'},\n",
       " {'ensembl_gene_id': 'ENSG00000243319', 'hgnc_symbol': 'FGF14-IT1'},\n",
       " {'ensembl_gene_id': 'ENSG00000243319', 'hgnc_symbol': 'FGF14-IT1'},\n",
       " {'ensembl_gene_id': 'ENSG00000224405', 'hgnc_symbol': 'LINC00572'},\n",
       " {'ensembl_gene_id': 'ENSG00000234445', 'hgnc_symbol': 'FGF14-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000234445', 'hgnc_symbol': 'FGF14-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000122043', 'hgnc_symbol': 'LINC00544'},\n",
       " {'ensembl_gene_id': 'ENSG00000122043', 'hgnc_symbol': 'LINC00544'},\n",
       " {'ensembl_gene_id': 'ENSG00000122043', 'hgnc_symbol': 'LINC00544'},\n",
       " {'ensembl_gene_id': 'ENSG00000272143', 'hgnc_symbol': 'FGF14-AS2'},\n",
       " {'ensembl_gene_id': 'ENSG00000266505', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000270581', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000261057', 'hgnc_symbol': 'LINC00555'},\n",
       " {'ensembl_gene_id': 'ENSG00000224511', 'hgnc_symbol': 'LINC00365'},\n",
       " {'ensembl_gene_id': 'ENSG00000224511', 'hgnc_symbol': 'LINC00365'},\n",
       " {'ensembl_gene_id': 'ENSG00000269942', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000232117', 'hgnc_symbol': 'LINC00384'},\n",
       " {'ensembl_gene_id': 'ENSG00000238869', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000231633', 'hgnc_symbol': 'LINC00283'},\n",
       " {'ensembl_gene_id': 'ENSG00000232643', 'hgnc_symbol': 'LINC00385'},\n",
       " {'ensembl_gene_id': 'ENSG00000222301', 'hgnc_symbol': 'RNY5P8'},\n",
       " {'ensembl_gene_id': 'ENSG00000266816', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000252928', 'hgnc_symbol': 'RNU6-64P'},\n",
       " {'ensembl_gene_id': 'ENSG00000250878', 'hgnc_symbol': 'METTL21EP'},\n",
       " {'ensembl_gene_id': 'ENSG00000250878', 'hgnc_symbol': 'METTL21EP'},\n",
       " {'ensembl_gene_id': 'ENSG00000215049', 'hgnc_symbol': 'PRDX2P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000272542', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000236463', 'hgnc_symbol': 'LINC00427'},\n",
       " {'ensembl_gene_id': 'ENSG00000234551', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000230474', 'hgnc_symbol': 'ATP6V1G1P7'},\n",
       " {'ensembl_gene_id': 'ENSG00000225823', 'hgnc_symbol': 'RPL7P45'},\n",
       " {'ensembl_gene_id': 'ENSG00000232307', 'hgnc_symbol': 'DAOA-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000238121', 'hgnc_symbol': 'LINC00426'},\n",
       " {'ensembl_gene_id': 'ENSG00000238121', 'hgnc_symbol': 'LINC00426'},\n",
       " {'ensembl_gene_id': 'ENSG00000238121', 'hgnc_symbol': 'LINC00426'},\n",
       " {'ensembl_gene_id': 'ENSG00000238121', 'hgnc_symbol': 'LINC00426'},\n",
       " {'ensembl_gene_id': 'ENSG00000238121', 'hgnc_symbol': 'LINC00426'},\n",
       " {'ensembl_gene_id': 'ENSG00000238121', 'hgnc_symbol': 'LINC00426'},\n",
       " {'ensembl_gene_id': 'ENSG00000238121', 'hgnc_symbol': 'LINC00426'},\n",
       " {'ensembl_gene_id': 'ENSG00000238121', 'hgnc_symbol': 'LINC00426'},\n",
       " {'ensembl_gene_id': 'ENSG00000238121', 'hgnc_symbol': 'LINC00426'},\n",
       " {'ensembl_gene_id': 'ENSG00000226620', 'hgnc_symbol': 'LINC00343'},\n",
       " {'ensembl_gene_id': 'ENSG00000226620', 'hgnc_symbol': 'LINC00343'},\n",
       " {'ensembl_gene_id': 'ENSG00000226620', 'hgnc_symbol': 'LINC00343'},\n",
       " {'ensembl_gene_id': 'ENSG00000226620', 'hgnc_symbol': 'LINC00343'},\n",
       " {'ensembl_gene_id': 'ENSG00000226620', 'hgnc_symbol': 'LINC00343'},\n",
       " {'ensembl_gene_id': 'ENSG00000226620', 'hgnc_symbol': 'LINC00343'},\n",
       " {'ensembl_gene_id': 'ENSG00000226620', 'hgnc_symbol': 'LINC00343'},\n",
       " {'ensembl_gene_id': 'ENSG00000226620', 'hgnc_symbol': 'LINC00343'},\n",
       " {'ensembl_gene_id': 'ENSG00000226620', 'hgnc_symbol': 'LINC00343'},\n",
       " {'ensembl_gene_id': 'ENSG00000226620', 'hgnc_symbol': 'LINC00343'},\n",
       " {'ensembl_gene_id': 'ENSG00000226620', 'hgnc_symbol': 'LINC00343'},\n",
       " {'ensembl_gene_id': 'ENSG00000226620', 'hgnc_symbol': 'LINC00343'},\n",
       " {'ensembl_gene_id': 'ENSG00000225039', 'hgnc_symbol': 'LINC01058'},\n",
       " {'ensembl_gene_id': 'ENSG00000261577', 'hgnc_symbol': 'JUP'},\n",
       " {'ensembl_gene_id': 'ENSG00000261577', 'hgnc_symbol': 'JUP'},\n",
       " {'ensembl_gene_id': 'ENSG00000261577', 'hgnc_symbol': 'JUP'},\n",
       " {'ensembl_gene_id': 'ENSG00000261577', 'hgnc_symbol': 'JUP'},\n",
       " {'ensembl_gene_id': 'ENSG00000261577', 'hgnc_symbol': 'JUP'},\n",
       " {'ensembl_gene_id': 'ENSG00000261577', 'hgnc_symbol': 'JUP'},\n",
       " {'ensembl_gene_id': 'ENSG00000261577', 'hgnc_symbol': 'JUP'},\n",
       " {'ensembl_gene_id': 'ENSG00000252550', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000236444', 'hgnc_symbol': 'UBE2L5P'},\n",
       " {'ensembl_gene_id': 'ENSG00000236444', 'hgnc_symbol': 'UBE2L5P'},\n",
       " {'ensembl_gene_id': 'ENSG00000222643', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000222682', 'hgnc_symbol': 'RNA5SP38'},\n",
       " {'ensembl_gene_id': 'ENSG00000213411', 'hgnc_symbol': 'RBM22P2'},\n",
       " {'ensembl_gene_id': 'ENSG00000233532', 'hgnc_symbol': 'LINC00460'},\n",
       " {'ensembl_gene_id': 'ENSG00000233532', 'hgnc_symbol': 'LINC00460'},\n",
       " {'ensembl_gene_id': 'ENSG00000232489', 'hgnc_symbol': 'MFAP1P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000174418', 'hgnc_symbol': 'RPL35P9'},\n",
       " {'ensembl_gene_id': 'ENSG00000224117', 'hgnc_symbol': 'PTPN2P2'},\n",
       " {'ensembl_gene_id': 'ENSG00000272274', 'hgnc_symbol': 'LINC00551'},\n",
       " {'ensembl_gene_id': 'ENSG00000230156', 'hgnc_symbol': 'LINC00443'},\n",
       " {'ensembl_gene_id': 'ENSG00000237879', 'hgnc_symbol': 'LINC00398'},\n",
       " {'ensembl_gene_id': 'ENSG00000227590', 'hgnc_symbol': 'ATP5G1P5'},\n",
       " {'ensembl_gene_id': 'ENSG00000174408', 'hgnc_symbol': 'PPIAP24'},\n",
       " {'ensembl_gene_id': 'ENSG00000224743', 'hgnc_symbol': 'TEX26-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000224743', 'hgnc_symbol': 'TEX26-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000224743', 'hgnc_symbol': 'TEX26-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000224743', 'hgnc_symbol': 'TEX26-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000224743', 'hgnc_symbol': 'TEX26-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000224743', 'hgnc_symbol': 'TEX26-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000224743', 'hgnc_symbol': 'TEX26-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000224743', 'hgnc_symbol': 'TEX26-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000224743', 'hgnc_symbol': 'TEX26-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000224743', 'hgnc_symbol': 'TEX26-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000224743', 'hgnc_symbol': 'TEX26-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000224743', 'hgnc_symbol': 'TEX26-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000224743', 'hgnc_symbol': 'TEX26-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000224743', 'hgnc_symbol': 'TEX26-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000224743', 'hgnc_symbol': 'TEX26-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000224743', 'hgnc_symbol': 'TEX26-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000224743', 'hgnc_symbol': 'TEX26-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000196777', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000239050', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000221316', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000201847', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000221650', 'hgnc_symbol': 'MIR1267'},\n",
       " {'ensembl_gene_id': 'ENSG00000221473', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000227248', 'hgnc_symbol': 'FAM155A-IT1'},\n",
       " {'ensembl_gene_id': 'ENSG00000223177', 'hgnc_symbol': 'RNA5SP39'},\n",
       " {'ensembl_gene_id': 'ENSG00000232087', 'hgnc_symbol': 'HCFC2P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000229938', 'hgnc_symbol': 'MYO16-AS2'},\n",
       " {'ensembl_gene_id': 'ENSG00000236094', 'hgnc_symbol': 'LINC00545'},\n",
       " {'ensembl_gene_id': 'ENSG00000236242', 'hgnc_symbol': 'MYO16-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000238185', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000223617', 'hgnc_symbol': 'LINC00370'},\n",
       " {'ensembl_gene_id': 'ENSG00000230233', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000230403', 'hgnc_symbol': 'LINC01066'},\n",
       " {'ensembl_gene_id': 'ENSG00000236053', 'hgnc_symbol': 'LINC01067'},\n",
       " {'ensembl_gene_id': 'ENSG00000261163', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000261163', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000261163', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000261163', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000265841', 'hgnc_symbol': 'MIR548X'},\n",
       " {'ensembl_gene_id': 'ENSG00000231894', 'hgnc_symbol': 'WDR95P'},\n",
       " {'ensembl_gene_id': 'ENSG00000229792', 'hgnc_symbol': 'LINC00399'},\n",
       " {'ensembl_gene_id': 'ENSG00000221041', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000234854', 'hgnc_symbol': 'LINC00676'},\n",
       " {'ensembl_gene_id': 'ENSG00000198618', 'hgnc_symbol': 'PPIAP22'},\n",
       " {'ensembl_gene_id': 'ENSG00000229427', 'hgnc_symbol': 'ANKRD26P4'},\n",
       " {'ensembl_gene_id': 'ENSG00000255813', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000201161', 'hgnc_symbol': 'RN7SKP10'},\n",
       " {'ensembl_gene_id': 'ENSG00000231428', 'hgnc_symbol': 'LINC00396'},\n",
       " {'ensembl_gene_id': 'ENSG00000232193', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000229715', 'hgnc_symbol': 'EEF1DP3'},\n",
       " {'ensembl_gene_id': 'ENSG00000229715', 'hgnc_symbol': 'EEF1DP3'},\n",
       " {'ensembl_gene_id': 'ENSG00000229715', 'hgnc_symbol': 'EEF1DP3'},\n",
       " {'ensembl_gene_id': 'ENSG00000226204', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000265885', 'hgnc_symbol': 'RN7SL783P'},\n",
       " {'ensembl_gene_id': 'ENSG00000238629', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000235965', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000224821', 'hgnc_symbol': 'COL4A2-AS2'},\n",
       " {'ensembl_gene_id': 'ENSG00000235822', 'hgnc_symbol': 'LINC01073'},\n",
       " {'ensembl_gene_id': 'ENSG00000229289', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000237637', 'hgnc_symbol': 'FRY-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000232814', 'hgnc_symbol': 'COL4A2-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000226818', 'hgnc_symbol': 'SLC6A6P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000230633', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000255831', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000231620', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000215515', 'hgnc_symbol': 'IFIT1P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000225239', 'hgnc_symbol': 'RPL21P107'},\n",
       " {'ensembl_gene_id': 'ENSG00000212609', 'hgnc_symbol': 'RNU1-139P'},\n",
       " {'ensembl_gene_id': 'ENSG00000259831', 'hgnc_symbol': 'LINC00567'},\n",
       " {'ensembl_gene_id': 'ENSG00000212293', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000230198', 'hgnc_symbol': 'RPL37P4'},\n",
       " {'ensembl_gene_id': 'ENSG00000270008', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000229152', 'hgnc_symbol': 'ANKRD10-IT1'},\n",
       " {'ensembl_gene_id': 'ENSG00000271486', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000229336', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000229800', 'hgnc_symbol': 'ATP8A2P2'},\n",
       " {'ensembl_gene_id': 'ENSG00000227105', 'hgnc_symbol': 'PARP1P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000215353', 'hgnc_symbol': 'C1QBPP'},\n",
       " {'ensembl_gene_id': 'ENSG00000233480', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000225760', 'hgnc_symbol': 'LINC00431'},\n",
       " {'ensembl_gene_id': 'ENSG00000233236', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000225870', 'hgnc_symbol': 'LINC00368'},\n",
       " {'ensembl_gene_id': 'ENSG00000251851', 'hgnc_symbol': 'RNU6-772P'},\n",
       " {'ensembl_gene_id': 'ENSG00000252963', 'hgnc_symbol': 'RN7SKP147'},\n",
       " {'ensembl_gene_id': 'ENSG00000235875', 'hgnc_symbol': 'ARHGEF7-AS2'},\n",
       " {'ensembl_gene_id': 'ENSG00000207325', 'hgnc_symbol': 'RNY1P4'},\n",
       " {'ensembl_gene_id': 'ENSG00000233676', 'hgnc_symbol': 'FDPSP6'},\n",
       " {'ensembl_gene_id': 'ENSG00000226968', 'hgnc_symbol': 'LINC00423'},\n",
       " {'ensembl_gene_id': 'ENSG00000233644', 'hgnc_symbol': 'ARHGEF7-IT1'},\n",
       " {'ensembl_gene_id': 'ENSG00000234439', 'hgnc_symbol': 'KRT18P2'},\n",
       " {'ensembl_gene_id': 'ENSG00000227352', 'hgnc_symbol': 'ARHGEF7-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000227604', 'hgnc_symbol': 'TOMM22P3'},\n",
       " {'ensembl_gene_id': 'ENSG00000233206', 'hgnc_symbol': 'RPS3AP1'},\n",
       " {'ensembl_gene_id': 'ENSG00000221677', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000215881', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000230300', 'hgnc_symbol': 'STARD13-IT1'},\n",
       " {'ensembl_gene_id': 'ENSG00000261258', 'hgnc_symbol': 'FKBP10'},\n",
       " {'ensembl_gene_id': 'ENSG00000261258', 'hgnc_symbol': 'FKBP10'},\n",
       " {'ensembl_gene_id': 'ENSG00000261258', 'hgnc_symbol': 'FKBP10'},\n",
       " {'ensembl_gene_id': 'ENSG00000261258', 'hgnc_symbol': 'FKBP10'},\n",
       " {'ensembl_gene_id': 'ENSG00000261258', 'hgnc_symbol': 'FKBP10'},\n",
       " {'ensembl_gene_id': 'ENSG00000261258', 'hgnc_symbol': 'FKBP10'},\n",
       " {'ensembl_gene_id': 'ENSG00000261258', 'hgnc_symbol': 'FKBP10'},\n",
       " {'ensembl_gene_id': 'ENSG00000261258', 'hgnc_symbol': 'FKBP10'},\n",
       " {'ensembl_gene_id': 'ENSG00000226903', 'hgnc_symbol': 'LINC00354'},\n",
       " {'ensembl_gene_id': 'ENSG00000226903', 'hgnc_symbol': 'LINC00354'},\n",
       " {'ensembl_gene_id': 'ENSG00000224924', 'hgnc_symbol': 'LINC00320'},\n",
       " {'ensembl_gene_id': 'ENSG00000224924', 'hgnc_symbol': 'LINC00320'},\n",
       " {'ensembl_gene_id': 'ENSG00000224924', 'hgnc_symbol': 'LINC00320'},\n",
       " {'ensembl_gene_id': 'ENSG00000224924', 'hgnc_symbol': 'LINC00320'},\n",
       " {'ensembl_gene_id': 'ENSG00000224924', 'hgnc_symbol': 'LINC00320'},\n",
       " {'ensembl_gene_id': 'ENSG00000264989', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000200072', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000260385', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000229520', 'hgnc_symbol': 'LINC00404'},\n",
       " {'ensembl_gene_id': 'ENSG00000215351', 'hgnc_symbol': 'PPIAP1'},\n",
       " {'ensembl_gene_id': 'ENSG00000224243', 'hgnc_symbol': 'LINC00403'},\n",
       " {'ensembl_gene_id': 'ENSG00000226771', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000260102', 'hgnc_symbol': 'LINC01070'},\n",
       " {'ensembl_gene_id': 'ENSG00000221210', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000234730', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000260343', 'hgnc_symbol': 'LINC01043'},\n",
       " {'ensembl_gene_id': 'ENSG00000198434', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000223626', 'hgnc_symbol': 'LINC01044'},\n",
       " {'ensembl_gene_id': 'ENSG00000237527', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000264726', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000232684', 'hgnc_symbol': 'ATP11A-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000238265', 'hgnc_symbol': 'LINC00317'},\n",
       " {'ensembl_gene_id': 'ENSG00000234603', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000236581', 'hgnc_symbol': 'STARD13-AS'},\n",
       " {'ensembl_gene_id': 'ENSG00000236581', 'hgnc_symbol': 'STARD13-AS'},\n",
       " {'ensembl_gene_id': 'ENSG00000236581', 'hgnc_symbol': 'STARD13-AS'},\n",
       " {'ensembl_gene_id': 'ENSG00000236581', 'hgnc_symbol': 'STARD13-AS'},\n",
       " {'ensembl_gene_id': 'ENSG00000236581', 'hgnc_symbol': 'STARD13-AS'},\n",
       " {'ensembl_gene_id': 'ENSG00000236581', 'hgnc_symbol': 'STARD13-AS'},\n",
       " {'ensembl_gene_id': 'ENSG00000236581', 'hgnc_symbol': 'STARD13-AS'},\n",
       " {'ensembl_gene_id': 'ENSG00000236581', 'hgnc_symbol': 'STARD13-AS'},\n",
       " {'ensembl_gene_id': 'ENSG00000236581', 'hgnc_symbol': 'STARD13-AS'},\n",
       " {'ensembl_gene_id': 'ENSG00000236581', 'hgnc_symbol': 'STARD13-AS'},\n",
       " {'ensembl_gene_id': 'ENSG00000236581', 'hgnc_symbol': 'STARD13-AS'},\n",
       " {'ensembl_gene_id': 'ENSG00000236581', 'hgnc_symbol': 'STARD13-AS'},\n",
       " {'ensembl_gene_id': 'ENSG00000236581', 'hgnc_symbol': 'STARD13-AS'},\n",
       " {'ensembl_gene_id': 'ENSG00000236581', 'hgnc_symbol': 'STARD13-AS'},\n",
       " {'ensembl_gene_id': 'ENSG00000236581', 'hgnc_symbol': 'STARD13-AS'},\n",
       " {'ensembl_gene_id': 'ENSG00000267868', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000188403', 'hgnc_symbol': 'IGHV1OR15-9'},\n",
       " {'ensembl_gene_id': 'ENSG00000235280', 'hgnc_symbol': 'MCF2L-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000269356', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000233997', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000233997', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000269376', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000231882', 'hgnc_symbol': 'F10-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000230371', 'hgnc_symbol': 'KARSP2'},\n",
       " {'ensembl_gene_id': 'ENSG00000259490', 'hgnc_symbol': 'IGHV3OR15-7'},\n",
       " {'ensembl_gene_id': 'ENSG00000269125', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000227075', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000232662', 'hgnc_symbol': 'LDHBP1'},\n",
       " {'ensembl_gene_id': 'ENSG00000270961', 'hgnc_symbol': 'IGHD5OR15-5A'},\n",
       " {'ensembl_gene_id': 'ENSG00000225083', 'hgnc_symbol': 'GRTP1-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000225083', 'hgnc_symbol': 'GRTP1-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000233215', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000233215', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000233215', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000233613', 'hgnc_symbol': 'DCUN1D2-AS'},\n",
       " {'ensembl_gene_id': 'ENSG00000233613', 'hgnc_symbol': 'DCUN1D2-AS'},\n",
       " {'ensembl_gene_id': 'ENSG00000271317', 'hgnc_symbol': 'IGHD4OR15-4A'},\n",
       " {'ensembl_gene_id': 'ENSG00000202347', 'hgnc_symbol': 'RNU1-16P'},\n",
       " {'ensembl_gene_id': 'ENSG00000227208', 'hgnc_symbol': 'LINC00552'},\n",
       " {'ensembl_gene_id': 'ENSG00000270783', 'hgnc_symbol': 'IGHD3OR15-3A'},\n",
       " {'ensembl_gene_id': 'ENSG00000233695', 'hgnc_symbol': 'GAS6-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000271197', 'hgnc_symbol': 'IGHD2OR15-2A'},\n",
       " {'ensembl_gene_id': 'ENSG00000272695', 'hgnc_symbol': 'GAS6-AS2'},\n",
       " {'ensembl_gene_id': 'ENSG00000184856', 'hgnc_symbol': 'LINC00308'},\n",
       " {'ensembl_gene_id': 'ENSG00000227911', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000237699', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000271336', 'hgnc_symbol': 'IGHD1OR15-1A'},\n",
       " {'ensembl_gene_id': 'ENSG00000264288', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000265259', 'hgnc_symbol': 'RN7SL247P'},\n",
       " {'ensembl_gene_id': 'ENSG00000260753', 'hgnc_symbol': 'NT5C3B'},\n",
       " {'ensembl_gene_id': 'ENSG00000260753', 'hgnc_symbol': 'NT5C3B'},\n",
       " {'ensembl_gene_id': 'ENSG00000260753', 'hgnc_symbol': 'NT5C3B'},\n",
       " {'ensembl_gene_id': 'ENSG00000260753', 'hgnc_symbol': 'NT5C3B'},\n",
       " {'ensembl_gene_id': 'ENSG00000260753', 'hgnc_symbol': 'NT5C3B'},\n",
       " {'ensembl_gene_id': 'ENSG00000260753', 'hgnc_symbol': 'NT5C3B'},\n",
       " {'ensembl_gene_id': 'ENSG00000260753', 'hgnc_symbol': 'NT5C3B'},\n",
       " {'ensembl_gene_id': 'ENSG00000260753', 'hgnc_symbol': 'NT5C3B'},\n",
       " {'ensembl_gene_id': 'ENSG00000260753', 'hgnc_symbol': 'NT5C3B'},\n",
       " {'ensembl_gene_id': 'ENSG00000260753', 'hgnc_symbol': 'NT5C3B'},\n",
       " {'ensembl_gene_id': 'ENSG00000260753', 'hgnc_symbol': 'NT5C3B'},\n",
       " {'ensembl_gene_id': 'ENSG00000260753', 'hgnc_symbol': 'NT5C3B'},\n",
       " {'ensembl_gene_id': 'ENSG00000260753', 'hgnc_symbol': 'NT5C3B'},\n",
       " {'ensembl_gene_id': 'ENSG00000260753', 'hgnc_symbol': 'NT5C3B'},\n",
       " {'ensembl_gene_id': 'ENSG00000260753', 'hgnc_symbol': 'NT5C3B'},\n",
       " {'ensembl_gene_id': 'ENSG00000260753', 'hgnc_symbol': 'NT5C3B'},\n",
       " {'ensembl_gene_id': 'ENSG00000234340', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000230490', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000230490', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000230490', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000212051', 'hgnc_symbol': 'MIR320C2'},\n",
       " {'ensembl_gene_id': 'ENSG00000226921', 'hgnc_symbol': 'LINC00454'},\n",
       " {'ensembl_gene_id': 'ENSG00000226921', 'hgnc_symbol': 'LINC00454'},\n",
       " {'ensembl_gene_id': 'ENSG00000226921', 'hgnc_symbol': 'LINC00454'},\n",
       " {'ensembl_gene_id': 'ENSG00000226921', 'hgnc_symbol': 'LINC00454'},\n",
       " {'ensembl_gene_id': 'ENSG00000226921', 'hgnc_symbol': 'LINC00454'},\n",
       " {'ensembl_gene_id': 'ENSG00000226921', 'hgnc_symbol': 'LINC00454'},\n",
       " {'ensembl_gene_id': 'ENSG00000226921', 'hgnc_symbol': 'LINC00454'},\n",
       " {'ensembl_gene_id': 'ENSG00000264365', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000273321', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000234535', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000232986', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000199636', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000265094', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000252397', 'hgnc_symbol': 'RNU5A-4P'},\n",
       " {'ensembl_gene_id': 'ENSG00000263663', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000227254', 'hgnc_symbol': 'VDAC1P12'},\n",
       " {'ensembl_gene_id': 'ENSG00000229373', 'hgnc_symbol': 'LINC00452'},\n",
       " {'ensembl_gene_id': 'ENSG00000229373', 'hgnc_symbol': 'LINC00452'},\n",
       " {'ensembl_gene_id': 'ENSG00000229373', 'hgnc_symbol': 'LINC00452'},\n",
       " {'ensembl_gene_id': 'ENSG00000199196', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000215405', 'hgnc_symbol': 'GOLGA6L6'},\n",
       " {'ensembl_gene_id': 'ENSG00000271850', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000225179', 'hgnc_symbol': 'LINC00457'},\n",
       " {'ensembl_gene_id': 'ENSG00000225179', 'hgnc_symbol': 'LINC00457'},\n",
       " {'ensembl_gene_id': 'ENSG00000231136', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000260910', 'hgnc_symbol': 'LINC00565'},\n",
       " {'ensembl_gene_id': 'ENSG00000271901', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000202339', 'hgnc_symbol': 'RNU4-45P'},\n",
       " {'ensembl_gene_id': 'ENSG00000232487', 'hgnc_symbol': 'RASA3-IT1'},\n",
       " {'ensembl_gene_id': 'ENSG00000226043', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000231910', 'hgnc_symbol': 'GAMTP2'},\n",
       " {'ensembl_gene_id': 'ENSG00000264539', 'hgnc_symbol': 'MIR548AR'},\n",
       " {'ensembl_gene_id': 'ENSG00000261728', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000223488', 'hgnc_symbol': 'MAPK6PS2'},\n",
       " {'ensembl_gene_id': 'ENSG00000265450', 'hgnc_symbol': 'MIR4502'},\n",
       " {'ensembl_gene_id': 'ENSG00000232795', 'hgnc_symbol': 'SCAND3P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000226158', 'hgnc_symbol': 'CLCP2'},\n",
       " {'ensembl_gene_id': 'ENSG00000234835', 'hgnc_symbol': 'PHBP13'},\n",
       " {'ensembl_gene_id': 'ENSG00000229723', 'hgnc_symbol': 'LINC01054'},\n",
       " {'ensembl_gene_id': 'ENSG00000268343', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000221519', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000236036', 'hgnc_symbol': 'LINC00445'},\n",
       " {'ensembl_gene_id': 'ENSG00000236036', 'hgnc_symbol': 'LINC00445'},\n",
       " {'ensembl_gene_id': 'ENSG00000207203', 'hgnc_symbol': 'RNU6-71P'},\n",
       " {'ensembl_gene_id': 'ENSG00000169861', 'hgnc_symbol': 'IGHV1OR21-1'},\n",
       " {'ensembl_gene_id': 'ENSG00000266047', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000223383', 'hgnc_symbol': 'HIST1H2APS6'},\n",
       " {'ensembl_gene_id': 'ENSG00000222602', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000235306', 'hgnc_symbol': 'GAPDHP34'},\n",
       " {'ensembl_gene_id': 'ENSG00000223543', 'hgnc_symbol': 'TCEAL4P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000215486', 'hgnc_symbol': 'ARL2BPP3'},\n",
       " {'ensembl_gene_id': 'ENSG00000215486', 'hgnc_symbol': 'ARL2BPP3'},\n",
       " {'ensembl_gene_id': 'ENSG00000261631', 'hgnc_symbol': 'KLHL10'},\n",
       " {'ensembl_gene_id': 'ENSG00000261631', 'hgnc_symbol': 'KLHL10'},\n",
       " {'ensembl_gene_id': 'ENSG00000261631', 'hgnc_symbol': 'KLHL10'},\n",
       " {'ensembl_gene_id': 'ENSG00000230031', 'hgnc_symbol': 'POTEB2'},\n",
       " {'ensembl_gene_id': 'ENSG00000271337', 'hgnc_symbol': 'NDE1P2'},\n",
       " {'ensembl_gene_id': 'ENSG00000236711', 'hgnc_symbol': 'SMAD9-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000228825', 'hgnc_symbol': 'LAMTOR3P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000232378', 'hgnc_symbol': 'RPL29P28'},\n",
       " {'ensembl_gene_id': 'ENSG00000226161', 'hgnc_symbol': 'EIF4A1P5'},\n",
       " {'ensembl_gene_id': 'ENSG00000221100', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000202395', 'hgnc_symbol': 'RN7SKP1'},\n",
       " {'ensembl_gene_id': 'ENSG00000230283', 'hgnc_symbol': 'RPS12P24'},\n",
       " {'ensembl_gene_id': 'ENSG00000234034', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000230390', 'hgnc_symbol': 'LINC01048'},\n",
       " {'ensembl_gene_id': 'ENSG00000263796', 'hgnc_symbol': 'RN7SL609P'},\n",
       " {'ensembl_gene_id': 'ENSG00000199276', 'hgnc_symbol': 'RNA5SP26'},\n",
       " {'ensembl_gene_id': 'ENSG00000225906', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000228036', 'hgnc_symbol': 'HSPD1P9'},\n",
       " {'ensembl_gene_id': 'ENSG00000230972', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000259896', 'hgnc_symbol': 'KLHL11'},\n",
       " {'ensembl_gene_id': 'ENSG00000223685', 'hgnc_symbol': 'LINC00571'},\n",
       " {'ensembl_gene_id': 'ENSG00000223685', 'hgnc_symbol': 'LINC00571'},\n",
       " {'ensembl_gene_id': 'ENSG00000223685', 'hgnc_symbol': 'LINC00571'},\n",
       " {'ensembl_gene_id': 'ENSG00000219368', 'hgnc_symbol': 'ZNF299P'},\n",
       " {'ensembl_gene_id': 'ENSG00000236354', 'hgnc_symbol': 'LINC00437'},\n",
       " {'ensembl_gene_id': 'ENSG00000231058', 'hgnc_symbol': 'MSANTD2P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000229437', 'hgnc_symbol': 'LINC00366'},\n",
       " {'ensembl_gene_id': 'ENSG00000223078', 'hgnc_symbol': 'RNU2-55P'},\n",
       " {'ensembl_gene_id': 'ENSG00000234701', 'hgnc_symbol': 'PRDX3P3'},\n",
       " {'ensembl_gene_id': 'ENSG00000228592', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000252795', 'hgnc_symbol': 'RNU6-56P'},\n",
       " {'ensembl_gene_id': 'ENSG00000212049', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000270824', 'hgnc_symbol': 'IGHD5OR15-5B'},\n",
       " {'ensembl_gene_id': 'ENSG00000225350', 'hgnc_symbol': 'FREM2-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000225221', 'hgnc_symbol': 'PLA2G12AP2'},\n",
       " {'ensembl_gene_id': 'ENSG00000270451', 'hgnc_symbol': 'IGHD4OR15-4B'},\n",
       " {'ensembl_gene_id': 'ENSG00000235550', 'hgnc_symbol': 'ANKRD26P2'},\n",
       " {'ensembl_gene_id': 'ENSG00000223822', 'hgnc_symbol': 'EEF1A1P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000271255', 'hgnc_symbol': 'IGHD3OR15-3B'},\n",
       " {'ensembl_gene_id': 'ENSG00000214823', 'hgnc_symbol': 'NXT1P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000271305', 'hgnc_symbol': 'IGHD2OR15-2B'},\n",
       " {'ensembl_gene_id': 'ENSG00000238408', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000216014', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000264171', 'hgnc_symbol': 'MIR4305'},\n",
       " {'ensembl_gene_id': 'ENSG00000237569', 'hgnc_symbol': 'TUBAP'},\n",
       " {'ensembl_gene_id': 'ENSG00000200526', 'hgnc_symbol': 'RNY4P14'},\n",
       " {'ensembl_gene_id': 'ENSG00000224351', 'hgnc_symbol': 'CDKN2AIPNLP3'},\n",
       " {'ensembl_gene_id': 'ENSG00000138593', 'hgnc_symbol': 'SECISBP2L'},\n",
       " {'ensembl_gene_id': 'ENSG00000138593', 'hgnc_symbol': 'SECISBP2L'},\n",
       " {'ensembl_gene_id': 'ENSG00000138593', 'hgnc_symbol': 'SECISBP2L'},\n",
       " {'ensembl_gene_id': 'ENSG00000138593', 'hgnc_symbol': 'SECISBP2L'},\n",
       " {'ensembl_gene_id': 'ENSG00000138593', 'hgnc_symbol': 'SECISBP2L'},\n",
       " {'ensembl_gene_id': 'ENSG00000138593', 'hgnc_symbol': 'SECISBP2L'},\n",
       " {'ensembl_gene_id': 'ENSG00000138593', 'hgnc_symbol': 'SECISBP2L'},\n",
       " {'ensembl_gene_id': 'ENSG00000138593', 'hgnc_symbol': 'SECISBP2L'},\n",
       " {'ensembl_gene_id': 'ENSG00000138593', 'hgnc_symbol': 'SECISBP2L'},\n",
       " {'ensembl_gene_id': 'ENSG00000138593', 'hgnc_symbol': 'SECISBP2L'},\n",
       " {'ensembl_gene_id': 'ENSG00000270444', 'hgnc_symbol': 'AZU1P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000212553', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000270185', 'hgnc_symbol': 'IGHD1OR15-1B'},\n",
       " {'ensembl_gene_id': 'ENSG00000230710', 'hgnc_symbol': 'LINC00332'},\n",
       " {'ensembl_gene_id': 'ENSG00000271429', 'hgnc_symbol': 'RPL17P51'},\n",
       " {'ensembl_gene_id': 'ENSG00000207458', 'hgnc_symbol': 'RNY3P9'},\n",
       " {'ensembl_gene_id': 'ENSG00000268531', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000230058', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000252812', 'hgnc_symbol': 'RN7SKP2'},\n",
       " {'ensembl_gene_id': 'ENSG00000215483', 'hgnc_symbol': 'LINC00598'},\n",
       " {'ensembl_gene_id': 'ENSG00000215483', 'hgnc_symbol': 'LINC00598'},\n",
       " {'ensembl_gene_id': 'ENSG00000215483', 'hgnc_symbol': 'LINC00598'},\n",
       " {'ensembl_gene_id': 'ENSG00000229456', 'hgnc_symbol': 'RLIMP1'},\n",
       " {'ensembl_gene_id': 'ENSG00000211491', 'hgnc_symbol': 'MIR320D1'},\n",
       " {'ensembl_gene_id': 'ENSG00000207652', 'hgnc_symbol': 'MIR621'},\n",
       " {'ensembl_gene_id': 'ENSG00000168852', 'hgnc_symbol': 'TPTE2P5'},\n",
       " {'ensembl_gene_id': 'ENSG00000168852', 'hgnc_symbol': 'TPTE2P5'},\n",
       " {'ensembl_gene_id': 'ENSG00000168852', 'hgnc_symbol': 'TPTE2P5'},\n",
       " {'ensembl_gene_id': 'ENSG00000168852', 'hgnc_symbol': 'TPTE2P5'},\n",
       " {'ensembl_gene_id': 'ENSG00000233917', 'hgnc_symbol': 'POTEB'},\n",
       " {'ensembl_gene_id': 'ENSG00000233917', 'hgnc_symbol': 'POTEB'},\n",
       " {'ensembl_gene_id': 'ENSG00000233917', 'hgnc_symbol': 'POTEB'},\n",
       " {'ensembl_gene_id': 'ENSG00000233917', 'hgnc_symbol': 'POTEB'},\n",
       " {'ensembl_gene_id': 'ENSG00000166157', 'hgnc_symbol': 'TPTE'},\n",
       " {'ensembl_gene_id': 'ENSG00000166157', 'hgnc_symbol': 'TPTE'},\n",
       " {'ensembl_gene_id': 'ENSG00000166157', 'hgnc_symbol': 'TPTE'},\n",
       " {'ensembl_gene_id': 'ENSG00000166157', 'hgnc_symbol': 'TPTE'},\n",
       " {'ensembl_gene_id': 'ENSG00000260245', 'hgnc_symbol': 'ACLY'},\n",
       " {'ensembl_gene_id': 'ENSG00000260245', 'hgnc_symbol': 'ACLY'},\n",
       " {'ensembl_gene_id': 'ENSG00000176268', 'hgnc_symbol': 'CYCSP34'},\n",
       " {'ensembl_gene_id': 'ENSG00000239827', 'hgnc_symbol': 'SUGT1P3'},\n",
       " {'ensembl_gene_id': 'ENSG00000239827', 'hgnc_symbol': 'SUGT1P3'},\n",
       " {'ensembl_gene_id': 'ENSG00000229473', 'hgnc_symbol': 'RGS17P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000264226', 'hgnc_symbol': 'MIR3168'},\n",
       " {'ensembl_gene_id': 'ENSG00000199698', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000244264', 'hgnc_symbol': 'RN7SL597P'},\n",
       " {'ensembl_gene_id': 'ENSG00000227716', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000215482', 'hgnc_symbol': 'CALM2P3'},\n",
       " {'ensembl_gene_id': 'ENSG00000231986', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000238651', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000225133', 'hgnc_symbol': 'MORF4L1P4'},\n",
       " {'ensembl_gene_id': 'ENSG00000235564', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000231404', 'hgnc_symbol': 'RAC1P3'},\n",
       " {'ensembl_gene_id': 'ENSG00000226996', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000264695', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000223280', 'hgnc_symbol': 'RNU6-57P'},\n",
       " {'ensembl_gene_id': 'ENSG00000224018', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000266489', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000214222', 'hgnc_symbol': 'TUBBP2'},\n",
       " {'ensembl_gene_id': 'ENSG00000224832', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000266181', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000205240', 'hgnc_symbol': 'OR7E36P'},\n",
       " {'ensembl_gene_id': 'ENSG00000237484', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000237484', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000237484', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000265485', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000265485', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000234655', 'hgnc_symbol': 'OR7E155P'},\n",
       " {'ensembl_gene_id': 'ENSG00000234655', 'hgnc_symbol': 'OR7E155P'},\n",
       " {'ensembl_gene_id': 'ENSG00000263553', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000266520', 'hgnc_symbol': 'RAC1P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000215480', 'hgnc_symbol': 'OR7E37P'},\n",
       " {'ensembl_gene_id': 'ENSG00000266809', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000263907', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000266573', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000266573', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000266573', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000263928', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000264042', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000264190', 'hgnc_symbol': 'MIR5006'},\n",
       " {'ensembl_gene_id': 'ENSG00000241406', 'hgnc_symbol': 'RN7SL515P'},\n",
       " {'ensembl_gene_id': 'ENSG00000206962', 'hgnc_symbol': 'RNU6-74P'},\n",
       " {'ensembl_gene_id': 'ENSG00000264345', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000264345', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000229696', 'hgnc_symbol': 'KARSP1'},\n",
       " {'ensembl_gene_id': 'ENSG00000229248', 'hgnc_symbol': 'WBP2P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000226093', 'hgnc_symbol': 'RPS28P8'},\n",
       " {'ensembl_gene_id': 'ENSG00000231530', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000265552', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000237263', 'hgnc_symbol': 'MAPK6PS3'},\n",
       " {'ensembl_gene_id': 'ENSG00000264434', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000265679', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000242216', 'hgnc_symbol': 'RN7SL97P'},\n",
       " {'ensembl_gene_id': 'ENSG00000270809', 'hgnc_symbol': 'CHCHD2P11'},\n",
       " {'ensembl_gene_id': 'ENSG00000264193', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000234213', 'hgnc_symbol': 'FHP1'},\n",
       " {'ensembl_gene_id': 'ENSG00000243925', 'hgnc_symbol': 'RPS24P18'},\n",
       " {'ensembl_gene_id': 'ENSG00000260703', 'hgnc_symbol': 'TTC25'},\n",
       " {'ensembl_gene_id': 'ENSG00000260703', 'hgnc_symbol': 'TTC25'},\n",
       " {'ensembl_gene_id': 'ENSG00000260703', 'hgnc_symbol': 'TTC25'},\n",
       " {'ensembl_gene_id': 'ENSG00000260703', 'hgnc_symbol': 'TTC25'},\n",
       " {'ensembl_gene_id': 'ENSG00000233259', 'hgnc_symbol': 'FABP3P2'},\n",
       " {'ensembl_gene_id': 'ENSG00000188985', 'hgnc_symbol': 'DHFRP1'},\n",
       " {'ensembl_gene_id': 'ENSG00000188985', 'hgnc_symbol': 'DHFRP1'},\n",
       " {'ensembl_gene_id': 'ENSG00000271216', 'hgnc_symbol': 'LINC01050'},\n",
       " {'ensembl_gene_id': 'ENSG00000229546', 'hgnc_symbol': 'LINC00428'},\n",
       " {'ensembl_gene_id': 'ENSG00000266227', 'hgnc_symbol': 'NPM1P2'},\n",
       " {'ensembl_gene_id': 'ENSG00000238025', 'hgnc_symbol': 'ZDHHC4P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000256715', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000263829', 'hgnc_symbol': 'FAM60BP'},\n",
       " {'ensembl_gene_id': 'ENSG00000229928', 'hgnc_symbol': 'LINC00400'},\n",
       " {'ensembl_gene_id': 'ENSG00000232540', 'hgnc_symbol': 'RPL36P19'},\n",
       " {'ensembl_gene_id': 'ENSG00000266237', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000238189', 'hgnc_symbol': 'ENOX1-AS2'},\n",
       " {'ensembl_gene_id': 'ENSG00000252921', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000233821', 'hgnc_symbol': 'ENOX1-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000264911', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000270522', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000207160', 'hgnc_symbol': 'RNU6-1289P'},\n",
       " {'ensembl_gene_id': 'ENSG00000263862', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000166351', 'hgnc_symbol': 'POTED'},\n",
       " {'ensembl_gene_id': 'ENSG00000252846', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000233725', 'hgnc_symbol': 'LINC00284'},\n",
       " {'ensembl_gene_id': 'ENSG00000233725', 'hgnc_symbol': 'LINC00284'},\n",
       " {'ensembl_gene_id': 'ENSG00000233725', 'hgnc_symbol': 'LINC00284'},\n",
       " {'ensembl_gene_id': 'ENSG00000233725', 'hgnc_symbol': 'LINC00284'},\n",
       " {'ensembl_gene_id': 'ENSG00000233725', 'hgnc_symbol': 'LINC00284'},\n",
       " {'ensembl_gene_id': 'ENSG00000233725', 'hgnc_symbol': 'LINC00284'},\n",
       " {'ensembl_gene_id': 'ENSG00000233725', 'hgnc_symbol': 'LINC00284'},\n",
       " {'ensembl_gene_id': 'ENSG00000263846', 'hgnc_symbol': 'CIAPIN1P'},\n",
       " {'ensembl_gene_id': 'ENSG00000260283', 'hgnc_symbol': 'CNP'},\n",
       " {'ensembl_gene_id': 'ENSG00000260283', 'hgnc_symbol': 'CNP'},\n",
       " {'ensembl_gene_id': 'ENSG00000260283', 'hgnc_symbol': 'CNP'},\n",
       " {'ensembl_gene_id': 'ENSG00000260283', 'hgnc_symbol': 'CNP'},\n",
       " {'ensembl_gene_id': 'ENSG00000260283', 'hgnc_symbol': 'CNP'},\n",
       " {'ensembl_gene_id': 'ENSG00000269011', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000260372', 'hgnc_symbol': 'AQP4-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000260372', 'hgnc_symbol': 'AQP4-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000260372', 'hgnc_symbol': 'AQP4-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000260372', 'hgnc_symbol': 'AQP4-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000179611', 'hgnc_symbol': 'DGKZP1'},\n",
       " {'ensembl_gene_id': 'ENSG00000168675', 'hgnc_symbol': 'LDLRAD4'},\n",
       " {'ensembl_gene_id': 'ENSG00000168675', 'hgnc_symbol': 'LDLRAD4'},\n",
       " {'ensembl_gene_id': 'ENSG00000168675', 'hgnc_symbol': 'LDLRAD4'},\n",
       " {'ensembl_gene_id': 'ENSG00000168675', 'hgnc_symbol': 'LDLRAD4'},\n",
       " {'ensembl_gene_id': 'ENSG00000168675', 'hgnc_symbol': 'LDLRAD4'},\n",
       " {'ensembl_gene_id': 'ENSG00000168675', 'hgnc_symbol': 'LDLRAD4'},\n",
       " {'ensembl_gene_id': 'ENSG00000168675', 'hgnc_symbol': 'LDLRAD4'},\n",
       " {'ensembl_gene_id': 'ENSG00000168675', 'hgnc_symbol': 'LDLRAD4'},\n",
       " {'ensembl_gene_id': 'ENSG00000168675', 'hgnc_symbol': 'LDLRAD4'},\n",
       " {'ensembl_gene_id': 'ENSG00000168675', 'hgnc_symbol': 'LDLRAD4'},\n",
       " {'ensembl_gene_id': 'ENSG00000168675', 'hgnc_symbol': 'LDLRAD4'},\n",
       " {'ensembl_gene_id': 'ENSG00000168675', 'hgnc_symbol': 'LDLRAD4'},\n",
       " {'ensembl_gene_id': 'ENSG00000168675', 'hgnc_symbol': 'LDLRAD4'},\n",
       " {'ensembl_gene_id': 'ENSG00000168675', 'hgnc_symbol': 'LDLRAD4'},\n",
       " {'ensembl_gene_id': 'ENSG00000168675', 'hgnc_symbol': 'LDLRAD4'},\n",
       " {'ensembl_gene_id': 'ENSG00000168675', 'hgnc_symbol': 'LDLRAD4'},\n",
       " {'ensembl_gene_id': 'ENSG00000168675', 'hgnc_symbol': 'LDLRAD4'},\n",
       " {'ensembl_gene_id': 'ENSG00000168675', 'hgnc_symbol': 'LDLRAD4'},\n",
       " {'ensembl_gene_id': 'ENSG00000226519', 'hgnc_symbol': 'LINC00390'},\n",
       " {'ensembl_gene_id': 'ENSG00000227258', 'hgnc_symbol': 'SMIM2-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000227258', 'hgnc_symbol': 'SMIM2-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000265369', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000265369', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000235285', 'hgnc_symbol': 'SMIM2-IT1'},\n",
       " {'ensembl_gene_id': 'ENSG00000263677', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000230731', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000263382', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000237361', 'hgnc_symbol': 'LINC01071'},\n",
       " {'ensembl_gene_id': 'ENSG00000237361', 'hgnc_symbol': 'LINC01071'},\n",
       " {'ensembl_gene_id': 'ENSG00000266184', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000266549', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000225727', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000225727', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000265374', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000262684', 'hgnc_symbol': 'DNAJC7'},\n",
       " {'ensembl_gene_id': 'ENSG00000262684', 'hgnc_symbol': 'DNAJC7'},\n",
       " {'ensembl_gene_id': 'ENSG00000230793', 'hgnc_symbol': 'SMARCE1P5'},\n",
       " {'ensembl_gene_id': 'ENSG00000188992', 'hgnc_symbol': 'LIPI'},\n",
       " {'ensembl_gene_id': 'ENSG00000188992', 'hgnc_symbol': 'LIPI'},\n",
       " {'ensembl_gene_id': 'ENSG00000188992', 'hgnc_symbol': 'LIPI'},\n",
       " {'ensembl_gene_id': 'ENSG00000238932', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000237585', 'hgnc_symbol': 'LINC00407'},\n",
       " {'ensembl_gene_id': 'ENSG00000235097', 'hgnc_symbol': 'LINC00330'},\n",
       " {'ensembl_gene_id': 'ENSG00000223732', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000265802', 'hgnc_symbol': 'RN7SL49P'},\n",
       " {'ensembl_gene_id': 'ENSG00000223341', 'hgnc_symbol': 'RN7SKP3'},\n",
       " {'ensembl_gene_id': 'ENSG00000271818', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000252055', 'hgnc_symbol': 'RNU6-69P'},\n",
       " {'ensembl_gene_id': 'ENSG00000270381', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000221163', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000253051', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000199477', 'hgnc_symbol': 'SNORA31'},\n",
       " {'ensembl_gene_id': 'ENSG00000273149', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000185272', 'hgnc_symbol': 'RBM11'},\n",
       " {'ensembl_gene_id': 'ENSG00000185272', 'hgnc_symbol': 'RBM11'},\n",
       " {'ensembl_gene_id': 'ENSG00000185272', 'hgnc_symbol': 'RBM11'},\n",
       " {'ensembl_gene_id': 'ENSG00000185272', 'hgnc_symbol': 'RBM11'},\n",
       " {'ensembl_gene_id': 'ENSG00000185272', 'hgnc_symbol': 'RBM11'},\n",
       " {'ensembl_gene_id': 'ENSG00000185272', 'hgnc_symbol': 'RBM11'},\n",
       " {'ensembl_gene_id': 'ENSG00000182974', 'hgnc_symbol': 'OR4M2'},\n",
       " {'ensembl_gene_id': 'ENSG00000226983', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000260770', 'hgnc_symbol': 'NKIRAS2'},\n",
       " {'ensembl_gene_id': 'ENSG00000260770', 'hgnc_symbol': 'NKIRAS2'},\n",
       " {'ensembl_gene_id': 'ENSG00000260770', 'hgnc_symbol': 'NKIRAS2'},\n",
       " {'ensembl_gene_id': 'ENSG00000260770', 'hgnc_symbol': 'NKIRAS2'},\n",
       " {'ensembl_gene_id': 'ENSG00000260770', 'hgnc_symbol': 'NKIRAS2'},\n",
       " {'ensembl_gene_id': 'ENSG00000260770', 'hgnc_symbol': 'NKIRAS2'},\n",
       " {'ensembl_gene_id': 'ENSG00000260770', 'hgnc_symbol': 'NKIRAS2'},\n",
       " {'ensembl_gene_id': 'ENSG00000260770', 'hgnc_symbol': 'NKIRAS2'},\n",
       " {'ensembl_gene_id': 'ENSG00000183706', 'hgnc_symbol': 'OR4N4'},\n",
       " {'ensembl_gene_id': 'ENSG00000170919', 'hgnc_symbol': 'TPT1-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000170919', 'hgnc_symbol': 'TPT1-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000170919', 'hgnc_symbol': 'TPT1-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000170919', 'hgnc_symbol': 'TPT1-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000170919', 'hgnc_symbol': 'TPT1-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000170919', 'hgnc_symbol': 'TPT1-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000170919', 'hgnc_symbol': 'TPT1-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000170919', 'hgnc_symbol': 'TPT1-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000170919', 'hgnc_symbol': 'TPT1-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000170919', 'hgnc_symbol': 'TPT1-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000170919', 'hgnc_symbol': 'TPT1-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000170919', 'hgnc_symbol': 'TPT1-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000170919', 'hgnc_symbol': 'TPT1-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000170919', 'hgnc_symbol': 'TPT1-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000170919', 'hgnc_symbol': 'TPT1-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000170919', 'hgnc_symbol': 'TPT1-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000170919', 'hgnc_symbol': 'TPT1-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000170919', 'hgnc_symbol': 'TPT1-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000170919', 'hgnc_symbol': 'TPT1-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000227090', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000230379', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000230379', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000230379', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000270505', 'hgnc_symbol': 'IGHV1OR15-1'},\n",
       " {'ensembl_gene_id': 'ENSG00000269182', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000259261', 'hgnc_symbol': 'IGHV4OR15-8'},\n",
       " {'ensembl_gene_id': 'ENSG00000237444', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000244278', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000155304', 'hgnc_symbol': 'HSPA13'},\n",
       " {'ensembl_gene_id': 'ENSG00000155304', 'hgnc_symbol': 'HSPA13'},\n",
       " {'ensembl_gene_id': 'ENSG00000155304', 'hgnc_symbol': 'HSPA13'},\n",
       " {'ensembl_gene_id': 'ENSG00000260320', 'hgnc_symbol': 'ZNF385C'},\n",
       " {'ensembl_gene_id': 'ENSG00000260320', 'hgnc_symbol': 'ZNF385C'},\n",
       " {'ensembl_gene_id': 'ENSG00000260320', 'hgnc_symbol': 'ZNF385C'},\n",
       " {'ensembl_gene_id': 'ENSG00000260320', 'hgnc_symbol': 'ZNF385C'},\n",
       " {'ensembl_gene_id': 'ENSG00000228886', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000214455', 'hgnc_symbol': 'RCN1P2'},\n",
       " {'ensembl_gene_id': 'ENSG00000214455', 'hgnc_symbol': 'RCN1P2'},\n",
       " {'ensembl_gene_id': 'ENSG00000251015', 'hgnc_symbol': 'SLC25A30-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000229303', 'hgnc_symbol': 'PPIAP25'},\n",
       " {'ensembl_gene_id': 'ENSG00000200709', 'hgnc_symbol': 'RNA5SP27'},\n",
       " {'ensembl_gene_id': 'ENSG00000197414', 'hgnc_symbol': 'GOLGA6L1'},\n",
       " {'ensembl_gene_id': 'ENSG00000197414', 'hgnc_symbol': 'GOLGA6L1'},\n",
       " {'ensembl_gene_id': 'ENSG00000232786', 'hgnc_symbol': 'TIMM9P3'},\n",
       " {'ensembl_gene_id': 'ENSG00000228869', 'hgnc_symbol': 'COX4I1P2'},\n",
       " {'ensembl_gene_id': 'ENSG00000235366', 'hgnc_symbol': 'LINC01055'},\n",
       " {'ensembl_gene_id': 'ENSG00000232512', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000232512', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000232512', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000237876', 'hgnc_symbol': 'AKR1B1P4'},\n",
       " {'ensembl_gene_id': 'ENSG00000260178', 'hgnc_symbol': 'DHX58'},\n",
       " {'ensembl_gene_id': 'ENSG00000260178', 'hgnc_symbol': 'DHX58'},\n",
       " {'ensembl_gene_id': 'ENSG00000260178', 'hgnc_symbol': 'DHX58'},\n",
       " {'ensembl_gene_id': 'ENSG00000223870', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000235903', 'hgnc_symbol': 'CPB2-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000235903', 'hgnc_symbol': 'CPB2-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000235903', 'hgnc_symbol': 'CPB2-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000222042', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000238627', 'hgnc_symbol': 'RNA5SP489'},\n",
       " {'ensembl_gene_id': 'ENSG00000240767', 'hgnc_symbol': 'RN7SL288P'},\n",
       " {'ensembl_gene_id': 'ENSG00000238483', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000252854', 'hgnc_symbol': 'RN7SKP5'},\n",
       " {'ensembl_gene_id': 'ENSG00000251106', 'hgnc_symbol': 'FAM206BP'},\n",
       " {'ensembl_gene_id': 'ENSG00000261097', 'hgnc_symbol': 'LINC00563'},\n",
       " {'ensembl_gene_id': 'ENSG00000223336', 'hgnc_symbol': 'RNU2-6P'},\n",
       " {'ensembl_gene_id': 'ENSG00000252385', 'hgnc_symbol': 'RNU6-68P'},\n",
       " {'ensembl_gene_id': 'ENSG00000166200', 'hgnc_symbol': 'COPS2'},\n",
       " {'ensembl_gene_id': 'ENSG00000166200', 'hgnc_symbol': 'COPS2'},\n",
       " {'ensembl_gene_id': 'ENSG00000166200', 'hgnc_symbol': 'COPS2'},\n",
       " {'ensembl_gene_id': 'ENSG00000166200', 'hgnc_symbol': 'COPS2'},\n",
       " {'ensembl_gene_id': 'ENSG00000166200', 'hgnc_symbol': 'COPS2'},\n",
       " {'ensembl_gene_id': 'ENSG00000166200', 'hgnc_symbol': 'COPS2'},\n",
       " {'ensembl_gene_id': 'ENSG00000155307', 'hgnc_symbol': 'SAMSN1'},\n",
       " {'ensembl_gene_id': 'ENSG00000155307', 'hgnc_symbol': 'SAMSN1'},\n",
       " {'ensembl_gene_id': 'ENSG00000155307', 'hgnc_symbol': 'SAMSN1'},\n",
       " {'ensembl_gene_id': 'ENSG00000155307', 'hgnc_symbol': 'SAMSN1'},\n",
       " {'ensembl_gene_id': 'ENSG00000155307', 'hgnc_symbol': 'SAMSN1'},\n",
       " {'ensembl_gene_id': 'ENSG00000241353', 'hgnc_symbol': 'PPP1R2P4'},\n",
       " {'ensembl_gene_id': 'ENSG00000241353', 'hgnc_symbol': 'PPP1R2P4'},\n",
       " {'ensembl_gene_id': 'ENSG00000259958', 'hgnc_symbol': 'KAT2A'},\n",
       " {'ensembl_gene_id': 'ENSG00000259958', 'hgnc_symbol': 'KAT2A'},\n",
       " {'ensembl_gene_id': 'ENSG00000243440', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000243440', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000243440', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000231817', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000231817', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000231817', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000231817', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000231817', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000231817', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000231817', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000231817', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000231817', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000227019', 'hgnc_symbol': 'OR7E101P'},\n",
       " {'ensembl_gene_id': 'ENSG00000205105', 'hgnc_symbol': 'COX17P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000205105', 'hgnc_symbol': 'COX17P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000260325', 'hgnc_symbol': 'HSPB9'},\n",
       " {'ensembl_gene_id': 'ENSG00000230243', 'hgnc_symbol': 'FKBP1AP3'},\n",
       " {'ensembl_gene_id': 'ENSG00000224517', 'hgnc_symbol': 'HTR2A-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000224517', 'hgnc_symbol': 'HTR2A-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000234590', 'hgnc_symbol': 'GNG5P5'},\n",
       " {'ensembl_gene_id': 'ENSG00000244521', 'hgnc_symbol': 'RN7SL700P'},\n",
       " {'ensembl_gene_id': 'ENSG00000234145', 'hgnc_symbol': 'NAP1L4P3'},\n",
       " {'ensembl_gene_id': 'ENSG00000229423', 'hgnc_symbol': 'RPL27AP8'},\n",
       " {'ensembl_gene_id': 'ENSG00000261378', 'hgnc_symbol': 'RAB5C'},\n",
       " {'ensembl_gene_id': 'ENSG00000261378', 'hgnc_symbol': 'RAB5C'},\n",
       " {'ensembl_gene_id': 'ENSG00000261378', 'hgnc_symbol': 'RAB5C'},\n",
       " {'ensembl_gene_id': 'ENSG00000261378', 'hgnc_symbol': 'RAB5C'},\n",
       " {'ensembl_gene_id': 'ENSG00000261378', 'hgnc_symbol': 'RAB5C'},\n",
       " {'ensembl_gene_id': 'ENSG00000261378', 'hgnc_symbol': 'RAB5C'},\n",
       " {'ensembl_gene_id': 'ENSG00000261378', 'hgnc_symbol': 'RAB5C'},\n",
       " {'ensembl_gene_id': 'ENSG00000228573', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000153575', 'hgnc_symbol': 'TUBGCP5'},\n",
       " {'ensembl_gene_id': 'ENSG00000153575', 'hgnc_symbol': 'TUBGCP5'},\n",
       " {'ensembl_gene_id': 'ENSG00000153575', 'hgnc_symbol': 'TUBGCP5'},\n",
       " {'ensembl_gene_id': 'ENSG00000153575', 'hgnc_symbol': 'TUBGCP5'},\n",
       " {'ensembl_gene_id': 'ENSG00000153575', 'hgnc_symbol': 'TUBGCP5'},\n",
       " {'ensembl_gene_id': 'ENSG00000153575', 'hgnc_symbol': 'TUBGCP5'},\n",
       " {'ensembl_gene_id': 'ENSG00000153575', 'hgnc_symbol': 'TUBGCP5'},\n",
       " {'ensembl_gene_id': 'ENSG00000234689', 'hgnc_symbol': 'LINC00444'},\n",
       " {'ensembl_gene_id': 'ENSG00000260388', 'hgnc_symbol': 'LINC00562'},\n",
       " {'ensembl_gene_id': 'ENSG00000227848', 'hgnc_symbol': 'SUCLA2-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000263172', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000229111', 'hgnc_symbol': 'MED4-AS1'},\n",
       " {'ensembl_gene_id': 'ENSG00000262688', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000266571', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000224510', 'hgnc_symbol': 'POLR2KP2'},\n",
       " {'ensembl_gene_id': 'ENSG00000260752', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000231473', 'hgnc_symbol': 'LINC00441'},\n",
       " {'ensembl_gene_id': 'ENSG00000231473', 'hgnc_symbol': 'LINC00441'},\n",
       " {'ensembl_gene_id': 'ENSG00000264151', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000264151', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000264151', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000264151', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000264151', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000238086', 'hgnc_symbol': 'PPP1R26P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000238086', 'hgnc_symbol': 'PPP1R26P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000177197', 'hgnc_symbol': 'PCNPP5'},\n",
       " {'ensembl_gene_id': 'ENSG00000233456', 'hgnc_symbol': 'LINC01077'},\n",
       " {'ensembl_gene_id': 'ENSG00000180530', 'hgnc_symbol': 'NRIP1'},\n",
       " {'ensembl_gene_id': 'ENSG00000180530', 'hgnc_symbol': 'NRIP1'},\n",
       " {'ensembl_gene_id': 'ENSG00000180530', 'hgnc_symbol': 'NRIP1'},\n",
       " {'ensembl_gene_id': 'ENSG00000180530', 'hgnc_symbol': 'NRIP1'},\n",
       " {'ensembl_gene_id': 'ENSG00000233610', 'hgnc_symbol': 'LINC00462'},\n",
       " {'ensembl_gene_id': 'ENSG00000264790', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000222391', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000225131', 'hgnc_symbol': 'PSME2P2'},\n",
       " {'ensembl_gene_id': 'ENSG00000225131', 'hgnc_symbol': 'PSME2P2'},\n",
       " {'ensembl_gene_id': 'ENSG00000244329', 'hgnc_symbol': 'UBA52P9'},\n",
       " {'ensembl_gene_id': 'ENSG00000265585', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000201662', 'hgnc_symbol': 'RNU6-60P'},\n",
       " {'ensembl_gene_id': 'ENSG00000263611', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000223724', 'hgnc_symbol': 'RAD17P2'},\n",
       " {'ensembl_gene_id': 'ENSG00000265097', 'hgnc_symbol': 'RBM22P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000199788', 'hgnc_symbol': 'RNY3P2'},\n",
       " {'ensembl_gene_id': 'ENSG00000235957', 'hgnc_symbol': 'COX7CP1'},\n",
       " {'ensembl_gene_id': 'ENSG00000263405', 'hgnc_symbol': 'PA2G4P3'},\n",
       " {'ensembl_gene_id': 'ENSG00000231665', 'hgnc_symbol': 'OGFOD1P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000227279', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000236577', 'hgnc_symbol': 'SNRPGP14'},\n",
       " {'ensembl_gene_id': 'ENSG00000137824', 'hgnc_symbol': 'RMDN3'},\n",
       " {'ensembl_gene_id': 'ENSG00000137824', 'hgnc_symbol': 'RMDN3'},\n",
       " {'ensembl_gene_id': 'ENSG00000137824', 'hgnc_symbol': 'RMDN3'},\n",
       " {'ensembl_gene_id': 'ENSG00000137824', 'hgnc_symbol': 'RMDN3'},\n",
       " {'ensembl_gene_id': 'ENSG00000137824', 'hgnc_symbol': 'RMDN3'},\n",
       " {'ensembl_gene_id': 'ENSG00000137824', 'hgnc_symbol': 'RMDN3'},\n",
       " {'ensembl_gene_id': 'ENSG00000137824', 'hgnc_symbol': 'RMDN3'},\n",
       " {'ensembl_gene_id': 'ENSG00000137824', 'hgnc_symbol': 'RMDN3'},\n",
       " {'ensembl_gene_id': 'ENSG00000137824', 'hgnc_symbol': 'RMDN3'},\n",
       " {'ensembl_gene_id': 'ENSG00000137824', 'hgnc_symbol': 'RMDN3'},\n",
       " {'ensembl_gene_id': 'ENSG00000137824', 'hgnc_symbol': 'RMDN3'},\n",
       " {'ensembl_gene_id': 'ENSG00000266296', 'hgnc_symbol': 'ARIH2P1'},\n",
       " {'ensembl_gene_id': 'ENSG00000176882', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000261371', 'hgnc_symbol': 'PECAM1'},\n",
       " {'ensembl_gene_id': 'ENSG00000261371', 'hgnc_symbol': 'PECAM1'},\n",
       " {'ensembl_gene_id': 'ENSG00000261371', 'hgnc_symbol': 'PECAM1'},\n",
       " {'ensembl_gene_id': 'ENSG00000261371', 'hgnc_symbol': 'PECAM1'},\n",
       " {'ensembl_gene_id': 'ENSG00000261371', 'hgnc_symbol': 'PECAM1'},\n",
       " {'ensembl_gene_id': 'ENSG00000261371', 'hgnc_symbol': 'PECAM1'},\n",
       " {'ensembl_gene_id': 'ENSG00000261371', 'hgnc_symbol': 'PECAM1'},\n",
       " {'ensembl_gene_id': 'ENSG00000261371', 'hgnc_symbol': 'PECAM1'},\n",
       " {'ensembl_gene_id': 'ENSG00000261371', 'hgnc_symbol': 'PECAM1'},\n",
       " {'ensembl_gene_id': 'ENSG00000265994', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000203547', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000271238', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000222148', 'hgnc_symbol': 'RNY4P30'},\n",
       " {'ensembl_gene_id': 'ENSG00000181358', 'hgnc_symbol': 'CTAGE10P'},\n",
       " {'ensembl_gene_id': 'ENSG00000251719', 'hgnc_symbol': 'RNU6-408P'},\n",
       " {'ensembl_gene_id': 'ENSG00000265730', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000200064', 'hgnc_symbol': 'RNY4P9'},\n",
       " {'ensembl_gene_id': 'ENSG00000212085', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000264526', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000264864', 'hgnc_symbol': 'MIR3613'},\n",
       " {'ensembl_gene_id': 'ENSG00000266479', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000259458', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000259458', 'hgnc_symbol': nan},\n",
       " {'ensembl_gene_id': 'ENSG00000265980', 'hgnc_symbol': nan},\n",
       " ...]"
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensembl_gene_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_relevant(gf, eg, prevalence=0.1, threshold = 10): # gf = expression matrix, eg = list of genes to exclude\n",
    "    '''\n",
    "    Outputs True if the gene is relevant for analysis.  We throw out excluded genes.\n",
    "    By default, we take genes that are found in at least 5% of all cells at a level of\n",
    "    10 counts or more.\n",
    "    '''\n",
    "    nonzero_count = (gf > threshold).sum(axis=0)\n",
    "    nonzero_proportion = nonzero_count / len(gf)\n",
    "    return [gene for gene in gf.columns if\n",
    "           (gene not in eg) & \n",
    "           (nonzero_proportion[gene] > prevalence)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(209, 203)"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bulks_raw.columns.get_loc('C1orf174'), bulks_raw.columns.get_loc('LRRC47')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_name</th>\n",
       "      <th>LRRC47</th>\n",
       "      <th>RN7SL574P</th>\n",
       "      <th>RP1-286D6.5</th>\n",
       "      <th>CEP104</th>\n",
       "      <th>DFFB</th>\n",
       "      <th>AL691523.1</th>\n",
       "      <th>C1orf174</th>\n",
       "      <th>LINC01134</th>\n",
       "      <th>RP13-614K11.1</th>\n",
       "      <th>RP13-614K11.2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P9855_2031</th>\n",
       "      <td>103</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>190</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P9855_2032</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>171</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P9855_2033</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P9855_2034</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>232</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P9855_2035</th>\n",
       "      <td>66</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>169</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P9855_2036</th>\n",
       "      <td>76</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>204</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_name   LRRC47  RN7SL574P  RP1-286D6.5  CEP104  DFFB  AL691523.1  \\\n",
       "P9855_2031     103          0            0       0     0           0   \n",
       "P9855_2032       0          0            0       0     0           0   \n",
       "P9855_2033       0          0            0       0     0           0   \n",
       "P9855_2034       0          0            0       2     0           0   \n",
       "P9855_2035      66          0            0       0     0           0   \n",
       "P9855_2036      76          0            0       0     0           0   \n",
       "\n",
       "gene_name   C1orf174  LINC01134  RP13-614K11.1  RP13-614K11.2  \n",
       "P9855_2031       190          0              0              0  \n",
       "P9855_2032       171          0              0              0  \n",
       "P9855_2033        50          0              0              0  \n",
       "P9855_2034       232          0              0              0  \n",
       "P9855_2035       169          0              0              0  \n",
       "P9855_2036       204          0              0              0  "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bulks_raw.sort_index().iloc[28:34,203:213]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\n",
    "genes_relevant = get_relevant(bulks_raw, exclude_genes)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "73 bulks are measured, from P9855_2019 to P9855_2031.\n",
      "63677 genes are measured, from ENSG00000223972 to ENSG00000235857.\n"
     ]
    },
    {
     "ename": "IndexError",
     "evalue": "list index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-96-2112b05f2a03>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"{} genes are measured, from {} to {}.\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgenes\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mgenes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mgenes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mgenes_excluded\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mgene\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mgene\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mexclude_genes\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mgene\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mgenes\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"{} TCR genes were excluded, from {} to {}.\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgenes_excluded\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgenes_excluded\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgenes_excluded\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"{} genes are considered relevant, from {} to {}.\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgenes_relevant\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgenes_relevant\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgenes_relevant\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mIndexError\u001b[0m: list index out of range"
     ]
    }
   ],
   "source": [
    "#print(\"{} cells are measured, from {} to {}.\".format(len(cells), cells[0], cells[-1]))\n",
    "print(\"{} bulks are measured, from {} to {}.\".format(len(bulks), bulks[0], bulks[-1]))\n",
    "print(\"{} genes are measured, from {} to {}.\".format(len(genes),genes[0],genes[-1]))\n",
    "genes_excluded = [gene for gene in exclude_genes if gene in genes]\n",
    "print(\"{} TCR genes were excluded, from {} to {}.\".format(len(genes_excluded), genes_excluded[0], genes_excluded[-1]))\n",
    "print(\"{} genes are considered relevant, from {} to {}.\".format(len(genes_relevant), genes_relevant[0], genes_relevant[-1]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Removing poor quality libraries and overexpressing cells\n",
    "\n",
    "Next we remove poor quality libraries -- by removing cells which express too few genes.  We also remove cells that may be dividing, and thus express too many genes.  To visualize this, we produce \"violin-plots\" giving the number of genes expressed by each cell, sorted by clonality."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def nGene(ge, md, cutoff = 10, plot=True): # ge = gene expression, md = meta data\n",
    "    nG = ge.apply(lambda row: sum(row > cutoff), axis=1) # Number of genes expressed.\n",
    "    nG.name = 'num_genes'\n",
    "    clonalities = md.Clone_ID\n",
    "    nGene_df = pd.concat([nG, clonalities], axis=1)\n",
    "    if plot:\n",
    "        fig,ax = plt.subplots(figsize=(12,8))\n",
    "        sns.violinplot(x=\"Clone_ID\", y=\"num_genes\", inner='quartiles', data=nGene_df, ax=ax)\n",
    "        sns.swarmplot(x=\"Clone_ID\", y=\"num_genes\", color=\"white\", size=3, data=nGene_df, ax=ax);\n",
    "    return nGene_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "nG = nGene(bulks_raw, meta_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we pick out the \"good cells\", whose expressed-gene-number is within two standard deviations of the mean (within each clone)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nG_std = nG.groupby('Clone_ID').apply(np.std)\n",
    "nG_means = nG.groupby('Clone_ID').apply(np.mean)\n",
    "nG_lowcut = (nG_means - 2*nG_std).num_genes\n",
    "nG_highcut = (nG_means + 2*nG_std).num_genes\n",
    "cells_good = [c for c in cells if \n",
    "              (nG.num_genes[c] >= nG_lowcut[nG.Clone_ID[c]]) and \n",
    "              (nG.num_genes[c] <= nG_highcut[nG.Clone_ID[c]]) \n",
    "             ]\n",
    "print('{} cells remaining after {} poor libraries removed.'.format(len(cells_good), len(cells) - len(cells_good)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "nG = nGene(bulks_raw.loc[cells_good], meta_df.loc[cells_good]) # Post-trimming violin-plot."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#  3.  MRN Normalization.\n",
    "\n",
    "Here we implement MRN Normalization on the gene expression data, closely following the convenient outline in Section 3.2 of [Maza](https://www.frontiersin.org/articles/10.3389/fgene.2016.00164/full).  We begin by putting our filtered data into a dataframe.  The dataframe `EM` (for \"expression matrix\") contains the gene expressions for the cells, filtered as above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "EM = bulks_raw#[genes_relevant]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "EM.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step I:  Prenormalization by library size.\n",
    "\n",
    "In Maza's article, $X_{gkr}$ stands for the raw count (number of reads) of gene $g$, for a cell number $r$ among clone $k$.  This information is contained in our expression matrix `EM`.  The first step is to normalize by library size, dividing $X_{gkr}$ by the total number of reads $N_{kr}$ of cell with numbers $k$, $r$.  \n",
    "\n",
    "We find the total number of reads for each cell by simply summing the numbers in each row of the data frame `EM`.  We don't need to worry about the separate indices $k$ and $r$ yet.  We examine the resulting \"library size\" below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "library_size = EM.sum(axis=1) # Drop the clone column.  Sum along rows.\n",
    "print(library_size.head())\n",
    "library_size.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The library size is about 1.6 million +/- 440,000.  Now we normalize the expression matrix by dividing every cell's raw counts by the cell's library size.  Note that `EM` is a dataframe whose rows are indexed by the cells.  `library_size` is a series (basically an array) whose rows are indexed by cells.  Numpy/pandas will divide one array by another, term by term, if they have the same size.  So it can divide *each column* of `EM` by `library size`, in a quickly-broadcasted division.  To perform this on every column, we use the `apply` method with a \"lambda\" function... it's the quickest method I know."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "Y = EM.apply(lambda column : column / library_size)\n",
    "Y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "Y.sort_index().iloc[28:31, 68:78] #just looking for some nonzero values in the df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now `Y` is the dataframe with counts normalized by library size, and we pass to the next step."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step II:  Creation of reference sample.\n",
    "\n",
    "A difference between TMM (used in edgeR), RLE (used in DeSeq2), and MRN, is how they create a reference sample.  In MRN, a reference sample is created by averaging the previous dataframe `Y` over cells within a single condition (clone).  We carry this out here.\n",
    "\n",
    "Note the method-chaining in defining `Y_clonal` below.  The `assign` method tacks on a new column to `Y` for the clone_IDs.  The `groupby` method then groups cells by clonotypes.  The `apply` method then takes the means within each clonotype.  The end result is to replace the (prenormalized) gene expression for *each cell* by the averages for each clonotype.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Y_clonal = Y.assign(Clone=meta_df['Clone_ID']).groupby('Clone').apply(np.mean) # Takes a few seconds.\n",
    "display(Y_clonal)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's take our reference sample to be the A7 clonal average."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Y_ref = Y_clonal.loc['D12'] # Our reference sample.  Basically an average of all cells of clonotype D12.\n",
    "Y_ref.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step III:  Computation of size relative to reference sample."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we compute relative scaling factors for each clone, based on the median fold-changes between gene expression within that clone and the reference clone (A7, we have chosen above).  Due to dropouts (values of 0 in gene expression), we discard genes with zeros when computing fold changes.  This avoids division-by-zero problems, and it matches the implementation of [Maza et al.](https://www.tandfonline.com/doi/full/10.4161/cib.25849).  (See the R code in the supplementary information).  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tau = pd.Series(index = clones)\n",
    "for clone in clones: # Why not a little loop.\n",
    "    numerator = Y_clonal.loc[clone]\n",
    "    denominator = Y_ref\n",
    "    ok_genes = [gene for gene in genes_relevant if (numerator[gene] != 0) and (denominator[gene] != 0)]\n",
    "    ratios = numerator[ok_genes] / denominator[ok_genes] # The ratios.\n",
    "    tau[clone] = np.median(ratios)\n",
    "print(tau)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step IV:  Adjustment of relative scaling factors."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This step does not occur in MRN normalization.  Yay!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step V:  Effective library size."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# intermediate series to match the clone IDs with the samples\n",
    "clonalities = pd.Series(meta_df['Clone_ID'], index=library_size.index)\n",
    "clonalities.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#intermediate series to map tau scaling values to the samples\n",
    "scalar_lib_size = clonalities.map(tau)\n",
    "scalar_lib_size.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#final series, for this step at least as the library_size values are scaled by tau values\n",
    "effective_lib_size = scalar_lib_size * library_size\n",
    "effective_lib_size.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step VI:  Computation of relative normalization/size factors."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "geo_mean = np.exp(effective_lib_size.apply(np.log).mean()) # The geometric mean of the effective_lib_size series.\n",
    "geo_mean # A single number!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "effective_lib_size.mean() # Compare the geometric mean (above) to the usual mean here.  They shouldn't be *too* far."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "relative_normalization_factor = effective_lib_size / geo_mean\n",
    "relative_normalization_factor.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step VII:  Normalization of counts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# here we normalize our raw counts by the relative normalization factor we just\n",
    "# calculated in Step VI\n",
    "mrn_counts = EM.div(relative_normalization_factor, axis=0)\n",
    "mrn_counts.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4.  Quick shot at machine learning\n",
    "\n",
    "After preprocessing the data, one might want to quickly check how well gene expression can predict clonotype.  The following steps will test out a linear support vector classifier. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import statsmodels.api as sm\n",
    "\n",
    "# Import various functions for stats.\n",
    "from scipy.stats import f_oneway, binom, binom_test, kruskal, norm # For p-values, kruskal-wallis, Gaussian\n",
    "from scipy.stats import kurtosis, skew # For detecting bimodality\n",
    "from statsmodels.formula.api import ols\n",
    "\n",
    "from bisect import bisect # See bisect.bisect.\n",
    "from itertools import compress # I don't recall why this is here.\n",
    "\n",
    "# Import from scikit, for machine learning.\n",
    "# from sklearn import mixture  # Import Gaussian mixture\n",
    "\n",
    "from sklearn.preprocessing import Binarizer, MinMaxScaler\n",
    "\n",
    "from sklearn.model_selection import ShuffleSplit, train_test_split, GridSearchCV\n",
    "\n",
    "from sklearn.feature_selection import SelectKBest, chi2, f_classif\n",
    "\n",
    "from sklearn.metrics import accuracy_score\n",
    "\n",
    "from sklearn.svm import LinearSVC # Linear Support Vector Classifier \n",
    "from sklearn.pipeline import make_pipeline\n",
    "from sklearn.pipeline import Pipeline\n",
    " \n",
    "from sklearn.cluster import MiniBatchKMeans, KMeans\n",
    "from sklearn.metrics.pairwise import pairwise_distances_argmin\n",
    "\n",
    "from sklearn.manifold import TSNE\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X = mrn_counts\n",
    "X_cells = mrn_counts.index\n",
    "y = meta_df.loc[X_cells].Clone_ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "SVMpipe = make_pipeline(MinMaxScaler(), \n",
    "                        SelectKBest(score_func = f_classif, k=100), \n",
    "                        LinearSVC(random_state=0, tol=1e-4, max_iter = 2000))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# DANGER DANGER:  The following suppresses all warnings from Python.\n",
    "# Otherwise, you'll probably get a bunch of convergence warnings...\n",
    "# Alternatively, increase the number of iterations in the SVM.\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore') # Ignore warnings.\n",
    "\n",
    "param_grid = dict(linearsvc__C=[0.001, 0.01, 0.1, 1.0, 10, 100],\n",
    "                  selectkbest__k = range(50,251,25),\n",
    "                 )\n",
    "\n",
    "grid_search = GridSearchCV(SVMpipe, param_grid=param_grid, cv=5, iid=False) #5-fold cross-validation.\n",
    "grid_search.fit(X,y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def CV_map(gs, swap=False):  # parameter_grid and grid_search.\n",
    "    '''\n",
    "    A general-purpose visualization for grid-search cross-validations.\n",
    "    Displays predictive accuracy as a heatmap, based on different\n",
    "    choices of parameters.  Displays optimal parameters in title.\n",
    "    \n",
    "    Args:\n",
    "        gs (sklearn.model_selection.GridSearchCV):  The grid-search object.\n",
    "        swap:  Set to True to switch the axes.\n",
    "        \n",
    "    Returns:\n",
    "        None\n",
    "        \n",
    "    '''\n",
    "    results = gs.cv_results_\n",
    "    means = results['mean_test_score']\n",
    "    params = list(gs.param_grid.keys())\n",
    "    \n",
    "    ser = pd.Series(list(means),\n",
    "                  index=pd.MultiIndex.from_tuples(\n",
    "                      [tuple(d.values()) for d in results['params']])\n",
    "               )\n",
    "    gs_df = ser.unstack().fillna(0)\n",
    "    gs_df.shape \n",
    "    if swap:\n",
    "        gs_df = gs_df.transpose()\n",
    "        params = params[::-1] # Switch the order of the parameter list.\n",
    "    \n",
    "    f, ax = plt.subplots(figsize=(12,12))\n",
    "    sns.heatmap(gs_df, annot=True, square=True, fmt=\"2.0%\", linewidths=.5, \n",
    "                cbar_kws={'fraction':0.025, 'pad':0.05}, ax=ax)\n",
    "    cbar = ax.collections[0].colorbar\n",
    "    cbar.set_ticks([.5, .6, .7, .8, .9])\n",
    "    cbar.set_ticklabels(['50%', '60%', '70%', '80%', '90%'])\n",
    "    ax.set_xlabel(params[1])\n",
    "    ax.set_ylabel(params[0])\n",
    "    plt.yticks(rotation=0)\n",
    "    bestacc = gs.best_score_\n",
    "    ax.set_title(\"Result of grid-search with 5-fold cross-validation. \\n\"\n",
    "                 \"Best score {0:2.1%} with parameters \\n\"\n",
    "                 \"{1} \".format(gs.best_score_, gs.best_params_))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "CV_map(grid_search)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn import base\n",
    "\n",
    "def confusion_matrix(X,y,pipe, reps=2, classlist = None):\n",
    "    if classlist:\n",
    "        classes = classlist\n",
    "    else:\n",
    "        classes = sorted(y.unique())\n",
    "    counts = pd.DataFrame(0, index=classes, columns=classes)\n",
    "    cpops = pd.DataFrame(0, index=classes, columns=['pop'])\n",
    "    correct = 0\n",
    "    total = 0\n",
    "    for t in range(reps):\n",
    "        this_pipe = base.clone(pipe)\n",
    "        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)\n",
    "        this_pipe.fit(X_train, y_train)\n",
    "        y_pred = this_pipe.predict(X_test)\n",
    "        actual = y_test\n",
    "        predicted = y_pred\n",
    "        for j in range(len(y_test)):\n",
    "            total += 1\n",
    "            if y_test[j] == y_pred[j]:\n",
    "                correct += 1\n",
    "            cpops.loc[y_test[j],'pop'] += 1\n",
    "            counts.loc[y_test[j], y_pred[j]] += 1\n",
    "    \n",
    "    valid = counts.apply(lambda column : column/cpops['pop']) #['pop'] )\n",
    "    \n",
    "    f, ax = plt.subplots(figsize=(len(classes)+1,len(classes)))\n",
    "    sns.heatmap(valid, annot=valid, square=True, fmt=\"2.0%\", linewidths=.5, \n",
    "                cbar_kws={'fraction':0.046, 'pad':0.04}, ax=ax)\n",
    "    cbar = ax.collections[0].colorbar\n",
    "    cbar.set_ticks([0, .25, .5, .75, 1])\n",
    "    cbar.set_ticklabels(['0%', '25%', '50%', '75%', '100%'])\n",
    "    ax.set_ylabel(\"Actual class\")\n",
    "    ax.set_xlabel(\"Predicted class\")\n",
    "    ax.set_title(\"Overall prediction accuracy {:0.3%}. \\n\"\n",
    "                 \"{} trials, in {} batches with independent splits.\".format(correct/total, total, reps))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "opt_k = grid_search.best_params_['selectkbest__k'] # Use optimal number of genes.\n",
    "opt_C = grid_search.best_params_['linearsvc__C'] # Use optimal C-parameter for linear SVM.\n",
    "\n",
    "SVMpipe = make_pipeline(MinMaxScaler(), \n",
    "                        SelectKBest(score_func = f_classif, k=opt_k), \n",
    "                        LinearSVC(random_state=0, tol=1e-4, max_iter = 2000, C=opt_C))\n",
    "\n",
    "confusion_matrix(X,y,SVMpipe, reps=50, classlist = clones)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
